// angular-archwizard/esm2015/lib/navigation/navigation-mode.interface.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * An interface describing the basic functionality, which must be provided by a navigation mode.
 * A navigation mode manages the navigation between different wizard steps, this contains the validation, if a step transition can be done
 *
 * @author Marc Arndt
 * @abstract
 */
export class NavigationMode {
    /**
     * @param {?} wizardState
     */
    constructor(wizardState) {
        this.wizardState = wizardState;
    }
    /**
     * Tries to transition the wizard to the previous step from the `currentStep`
     * @param {?=} preFinalize
     * @param {?=} postFinalize
     * @return {?}
     */
    goToPreviousStep(preFinalize, postFinalize) {
        if (this.wizardState.hasPreviousStep()) {
            this.goToStep(this.wizardState.currentStepIndex - 1, preFinalize, postFinalize);
        }
    }
    /**
     * Tries to transition the wizard to the next step from the `currentStep`
     * @param {?=} preFinalize
     * @param {?=} postFinalize
     * @return {?}
     */
    goToNextStep(preFinalize, postFinalize) {
        if (this.wizardState.hasNextStep()) {
            this.goToStep(this.wizardState.currentStepIndex + 1, preFinalize, postFinalize);
        }
    }
}
if (false) {
    /**
     * @type {?}
     * @protected
     */
    NavigationMode.prototype.wizardState;
    /**
     * Checks, whether a wizard step, as defined by the given destination index, can be transitioned to.
     *
     * @abstract
     * @param {?} destinationIndex The index of the destination step
     * @return {?} A [[Promise]] containing `true`, if the destination step can be transitioned to and false otherwise
     */
    NavigationMode.prototype.canGoToStep = function (destinationIndex) { };
    /**
     * Tries to transition to the wizard step, as denoted by the given destination index.
     * If this is not possible, the current wizard step should be exited and then reentered with `MovingDirection.Stay`
     *
     * @abstract
     * @param {?} destinationIndex The index of the destination step
     * @param {?=} preFinalize An event emitter, to be called before the step has been transitioned
     * @param {?=} postFinalize An event emitter, to be called after the step has been transitioned
     * @return {?}
     */
    NavigationMode.prototype.goToStep = function (destinationIndex, preFinalize, postFinalize) { };
    /**
     * Checks, whether the wizard step, located at the given index, is can be navigated to
     *
     * @abstract
     * @param {?} destinationIndex The index of the destination step
     * @return {?} True if the step can be navigated to, false otherwise
     */
    NavigationMode.prototype.isNavigable = function (destinationIndex) { };
    /**
     * Resets the state of this wizard.
     * A reset transitions the wizard automatically to the first step and sets all steps as incomplete.
     * In addition the whole wizard is set as incomplete
     * @abstract
     * @return {?}
     */
    NavigationMode.prototype.reset = function () { };
}
