// angular-archwizard/bundles/angular-archwizard.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('angular-archwizard', ['exports', '@angular/core', '@angular/common'], factory) :
    (factory((global['angular-archwizard'] = {}),global.ng.core,global.ng.common));
}(this, (function (exports,core,common) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    /* global Reflect, Promise */
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b)
                if (b.hasOwnProperty(p))
                    d[p] = b[p]; };
        return extendStatics(d, b);
    };
    function __extends(d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    }
    function __values(o) {
        var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
        if (m)
            return m.call(o);
        return {
            next: function () {
                if (o && i >= o.length)
                    o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `awWizardStepSymbol` directive can be used as an alternative to the `navigationSymbol` input of a [[WizardStep]]
     * to define the step symbol inside the navigation bar.  This way step symbol may contain arbitrary content.
     *
     * ### Syntax
     *
     * ```html
     * <ng-template awWizardStepSymbol>
     *     ...
     * </ng-template>
     * ```
     */
    var WizardStepSymbolDirective = /** @class */ (function () {
        /**
         * Constructor
         *
         * @param templateRef A reference to the content of the `ng-template` that contains this [[WizardStepSymbolDirective]]
         */
        function WizardStepSymbolDirective(templateRef) {
            this.templateRef = templateRef;
        }
        WizardStepSymbolDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: 'ng-template[awStepSymbol], ng-template[awWizardStepSymbol]'
                    },] }
        ];
        /** @nocollapse */
        WizardStepSymbolDirective.ctorParameters = function () {
            return [
                { type: core.TemplateRef }
            ];
        };
        return WizardStepSymbolDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `awWizardStepTitle` directive can be used as an alternative to the `stepTitle` input of a [[WizardStep]]
     * to define the content of a step title inside the navigation bar.
     * This step title can be freely created and can contain more than only plain text
     *
     * ### Syntax
     *
     * ```html
     * <ng-template awWizardStepTitle>
     *     ...
     * </ng-template>
     * ```
     *
     * @author Marc Arndt
     */
    var WizardStepTitleDirective = /** @class */ (function () {
        /**
         * Constructor
         *
         * @param templateRef A reference to the content of the `ng-template` that contains this [[WizardStepTitleDirective]]
         */
        function WizardStepTitleDirective(templateRef) {
            this.templateRef = templateRef;
        }
        WizardStepTitleDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: 'ng-template[awStepTitle], ng-template[awWizardStepTitle]'
                    },] }
        ];
        /** @nocollapse */
        WizardStepTitleDirective.ctorParameters = function () {
            return [
                { type: core.TemplateRef }
            ];
        };
        return WizardStepTitleDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * Basic functionality every type of wizard step needs to provide
     *
     * @author Marc Arndt
     * @abstract
     */
    var WizardStep = /** @class */ (function () {
        function WizardStep() {
            /**
             * A symbol property, which contains an optional symbol for the step inside the navigation bar.
             * Takes effect when `stepSymbolTemplate` is not defined or null.
             */
            this.navigationSymbol = { symbol: '' };
            /**
             * A boolean describing if the wizard step has been completed
             */
            this.completed = false;
            /**
             * A boolean describing if the wizard step is currently selected
             */
            this.selected = false;
            /**
             * A boolean describing, if the wizard step should be selected by default, i.e. after the wizard has been initialized as the initial step
             */
            this.defaultSelected = false;
            /**
             * A boolean describing if the wizard step is an optional step
             */
            this.optional = false;
            /**
             * A function or boolean deciding, if this step can be entered
             */
            this.canEnter = true;
            /**
             * A function or boolean deciding, if this step can be exited
             */
            this.canExit = true;
            /**
             * This [[EventEmitter]] is called when the step is entered.
             * The bound method should be used to do initialization work.
             */
            this.stepEnter = new core.EventEmitter();
            /**
             * This [[EventEmitter]] is called when the step is exited.
             * The bound method can be used to do cleanup work.
             */
            this.stepExit = new core.EventEmitter();
        }
        Object.defineProperty(WizardStep.prototype, "hidden", {
            /**
             * Returns if this wizard step should be visible to the user.
             * If the step should be visible to the user false is returned, otherwise true
             */
            get: /**
             * Returns if this wizard step should be visible to the user.
             * If the step should be visible to the user false is returned, otherwise true
             * @return {?}
             */ function () {
                return !this.selected;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * This method returns true, if this wizard step can be transitioned with a given direction.
         * Transitioned in this case means either entered or exited, depending on the given `condition` parameter.
         *
         * @param condition A condition variable, deciding if the step can be transitioned
         * @param direction The direction in which this step should be transitioned
         * @returns A [[Promise]] containing `true`, if this step can transitioned in the given direction
         * @throws An `Error` is thrown if `condition` is neither a function nor a boolean
         */
        /**
         * This method returns true, if this wizard step can be transitioned with a given direction.
         * Transitioned in this case means either entered or exited, depending on the given `condition` parameter.
         *
         * @throws An `Error` is thrown if `condition` is neither a function nor a boolean
         * @private
         * @param {?} condition A condition variable, deciding if the step can be transitioned
         * @param {?} direction The direction in which this step should be transitioned
         * @return {?} A [[Promise]] containing `true`, if this step can transitioned in the given direction
         */
        WizardStep.canTransitionStep = /**
         * This method returns true, if this wizard step can be transitioned with a given direction.
         * Transitioned in this case means either entered or exited, depending on the given `condition` parameter.
         *
         * @throws An `Error` is thrown if `condition` is neither a function nor a boolean
         * @private
         * @param {?} condition A condition variable, deciding if the step can be transitioned
         * @param {?} direction The direction in which this step should be transitioned
         * @return {?} A [[Promise]] containing `true`, if this step can transitioned in the given direction
         */
            function (condition, direction) {
                if (typeof (condition) === typeof (true)) {
                    return Promise.resolve(( /** @type {?} */(condition)));
                }
                else if (condition instanceof Function) {
                    return Promise.resolve(condition(direction));
                }
                else {
                    return Promise.reject(new Error("Input value '" + condition + "' is neither a boolean nor a function"));
                }
            };
        /**
         * A function called when the step is entered
         *
         * @param direction The direction in which the step is entered
         */
        /**
         * A function called when the step is entered
         *
         * @param {?} direction The direction in which the step is entered
         * @return {?}
         */
        WizardStep.prototype.enter = /**
         * A function called when the step is entered
         *
         * @param {?} direction The direction in which the step is entered
         * @return {?}
         */
            function (direction) {
                this.stepEnter.emit(direction);
            };
        /**
         * A function called when the step is exited
         *
         * @param direction The direction in which the step is exited
         */
        /**
         * A function called when the step is exited
         *
         * @param {?} direction The direction in which the step is exited
         * @return {?}
         */
        WizardStep.prototype.exit = /**
         * A function called when the step is exited
         *
         * @param {?} direction The direction in which the step is exited
         * @return {?}
         */
            function (direction) {
                this.stepExit.emit(direction);
            };
        /**
         * This method returns true, if this wizard step can be entered from the given direction.
         * Because this method depends on the value `canEnter`, it will throw an error, if `canEnter` is neither a boolean
         * nor a function.
         *
         * @param direction The direction in which this step should be entered
         * @returns A [[Promise]] containing `true`, if the step can be entered in the given direction, false otherwise
         * @throws An `Error` is thrown if `anEnter` is neither a function nor a boolean
         */
        /**
         * This method returns true, if this wizard step can be entered from the given direction.
         * Because this method depends on the value `canEnter`, it will throw an error, if `canEnter` is neither a boolean
         * nor a function.
         *
         * @throws An `Error` is thrown if `anEnter` is neither a function nor a boolean
         * @param {?} direction The direction in which this step should be entered
         * @return {?} A [[Promise]] containing `true`, if the step can be entered in the given direction, false otherwise
         */
        WizardStep.prototype.canEnterStep = /**
         * This method returns true, if this wizard step can be entered from the given direction.
         * Because this method depends on the value `canEnter`, it will throw an error, if `canEnter` is neither a boolean
         * nor a function.
         *
         * @throws An `Error` is thrown if `anEnter` is neither a function nor a boolean
         * @param {?} direction The direction in which this step should be entered
         * @return {?} A [[Promise]] containing `true`, if the step can be entered in the given direction, false otherwise
         */
            function (direction) {
                return WizardStep.canTransitionStep(this.canEnter, direction);
            };
        /**
         * This method returns true, if this wizard step can be exited into given direction.
         * Because this method depends on the value `canExit`, it will throw an error, if `canExit` is neither a boolean
         * nor a function.
         *
         * @param direction The direction in which this step should be left
         * @returns A [[Promise]] containing `true`, if the step can be exited in the given direction, false otherwise
         * @throws An `Error` is thrown if `canExit` is neither a function nor a boolean
         */
        /**
         * This method returns true, if this wizard step can be exited into given direction.
         * Because this method depends on the value `canExit`, it will throw an error, if `canExit` is neither a boolean
         * nor a function.
         *
         * @throws An `Error` is thrown if `canExit` is neither a function nor a boolean
         * @param {?} direction The direction in which this step should be left
         * @return {?} A [[Promise]] containing `true`, if the step can be exited in the given direction, false otherwise
         */
        WizardStep.prototype.canExitStep = /**
         * This method returns true, if this wizard step can be exited into given direction.
         * Because this method depends on the value `canExit`, it will throw an error, if `canExit` is neither a boolean
         * nor a function.
         *
         * @throws An `Error` is thrown if `canExit` is neither a function nor a boolean
         * @param {?} direction The direction in which this step should be left
         * @return {?} A [[Promise]] containing `true`, if the step can be exited in the given direction, false otherwise
         */
            function (direction) {
                return WizardStep.canTransitionStep(this.canExit, direction);
            };
        WizardStep.propDecorators = {
            stepTitleTemplate: [{ type: core.ContentChild, args: [WizardStepTitleDirective,] }],
            stepSymbolTemplate: [{ type: core.ContentChild, args: [WizardStepSymbolDirective,] }],
            stepId: [{ type: core.Input }],
            stepTitle: [{ type: core.Input }],
            navigationSymbol: [{ type: core.Input }],
            canEnter: [{ type: core.Input }],
            canExit: [{ type: core.Input }],
            stepEnter: [{ type: core.Output }],
            stepExit: [{ type: core.Output }],
            hidden: [{ type: core.HostBinding, args: ['hidden',] }]
        };
        return WizardStep;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * Basic functionality every wizard completion step needs to provide
     *
     * @author Marc Arndt
     * @abstract
     */
    var /**
     * Basic functionality every wizard completion step needs to provide
     *
     * @author Marc Arndt
     * @abstract
     */ WizardCompletionStep = /** @class */ (function (_super) {
        __extends(WizardCompletionStep, _super);
        function WizardCompletionStep() {
            var _this = _super !== null && _super.apply(this, arguments) || this;
            /**
             * @inheritDoc
             */
            _this.stepExit = new core.EventEmitter();
            /**
             * @inheritDoc
             */
            _this.canExit = false;
            return _this;
        }
        /**
         * @inheritDoc
         */
        /**
         * @inheritDoc
         * @param {?} direction
         * @return {?}
         */
        WizardCompletionStep.prototype.enter = /**
         * @inheritDoc
         * @param {?} direction
         * @return {?}
         */
            function (direction) {
                this.completed = true;
                this.stepEnter.emit(direction);
            };
        /**
         * @inheritDoc
         */
        /**
         * @inheritDoc
         * @param {?} direction
         * @return {?}
         */
        WizardCompletionStep.prototype.exit = /**
         * @inheritDoc
         * @param {?} direction
         * @return {?}
         */
            function (direction) {
                // set this completion step as incomplete
                this.completed = false;
                this.stepExit.emit(direction);
            };
        return WizardCompletionStep;
    }(WizardStep));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `aw-wizard-completion-step` component can be used to define a completion/success step at the end of your wizard
     * After a `aw-wizard-completion-step` has been entered, it has the characteristic that the user is blocked from
     * leaving it again to a previous step.
     * In addition entering a `aw-wizard-completion-step` automatically sets the `aw-wizard` and all steps inside the `aw-wizard`
     * as completed.
     *
     * ### Syntax
     *
     * ```html
     * <aw-wizard-completion-step [stepTitle]="title of the wizard step"
     *    [navigationSymbol]="{ symbol: 'navigation symbol', fontFamily: 'navigation symbol font family' }"
     *    (stepEnter)="event emitter to be called when the wizard step is entered"
     *    (stepExit)="event emitter to be called when the wizard step is exited">
     *    ...
     * </aw-wizard-completion-step>
     * ```
     *
     * ### Example
     *
     * ```html
     * <aw-wizard-completion-step stepTitle="Step 1" [navigationSymbol]="{ symbol: '1' }">
     *    ...
     * </aw-wizard-completion-step>
     * ```
     *
     * With a navigation symbol from the `font-awesome` font:
     *
     * ```html
     * <aw-wizard-completion-step stepTitle="Step 1" [navigationSymbol]="{ symbol: '&#xf1ba;', fontFamily: 'FontAwesome' }">
     *    ...
     * </aw-wizard-completion-step>
     * ```
     *
     * @author Marc Arndt
     */
    var WizardCompletionStepComponent = /** @class */ (function (_super) {
        __extends(WizardCompletionStepComponent, _super);
        function WizardCompletionStepComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        WizardCompletionStepComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'aw-wizard-completion-step',
                        template: "<ng-content></ng-content>\n",
                        encapsulation: core.ViewEncapsulation.None,
                        providers: [
                            { provide: WizardStep, useExisting: core.forwardRef(( /**
                                     * @return {?}
                                     */function () { return WizardCompletionStepComponent; })) },
                            { provide: WizardCompletionStep, useExisting: core.forwardRef(( /**
                                     * @return {?}
                                     */function () { return WizardCompletionStepComponent; })) }
                        ],
                        styles: ["aw-wizard-completion-step{height:auto;width:100%}"]
                    }] }
        ];
        return WizardCompletionStepComponent;
    }(WizardCompletionStep));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The direction in which a step transition was made
     *
     * @author Marc Arndt
     */
    /** @enum {number} */
    var MovingDirection = {
        /**
         * A forward step transition
         */
        Forwards: 0,
        /**
         * A backward step transition
         */
        Backwards: 1,
        /**
         * No step transition was done
         */
        Stay: 2,
    };
    MovingDirection[MovingDirection.Forwards] = 'Forwards';
    MovingDirection[MovingDirection.Backwards] = 'Backwards';
    MovingDirection[MovingDirection.Stay] = 'Stay';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * An interface describing the basic functionality, which must be provided by a navigation mode.
     * A navigation mode manages the navigation between different wizard steps, this contains the validation, if a step transition can be done
     *
     * @author Marc Arndt
     * @abstract
     */
    var /**
     * An interface describing the basic functionality, which must be provided by a navigation mode.
     * A navigation mode manages the navigation between different wizard steps, this contains the validation, if a step transition can be done
     *
     * @author Marc Arndt
     * @abstract
     */ NavigationMode = /** @class */ (function () {
        function NavigationMode(wizardState) {
            this.wizardState = wizardState;
        }
        /**
         * Tries to transition the wizard to the previous step from the `currentStep`
         */
        /**
         * Tries to transition the wizard to the previous step from the `currentStep`
         * @param {?=} preFinalize
         * @param {?=} postFinalize
         * @return {?}
         */
        NavigationMode.prototype.goToPreviousStep = /**
         * Tries to transition the wizard to the previous step from the `currentStep`
         * @param {?=} preFinalize
         * @param {?=} postFinalize
         * @return {?}
         */
            function (preFinalize, postFinalize) {
                if (this.wizardState.hasPreviousStep()) {
                    this.goToStep(this.wizardState.currentStepIndex - 1, preFinalize, postFinalize);
                }
            };
        /**
         * Tries to transition the wizard to the next step from the `currentStep`
         */
        /**
         * Tries to transition the wizard to the next step from the `currentStep`
         * @param {?=} preFinalize
         * @param {?=} postFinalize
         * @return {?}
         */
        NavigationMode.prototype.goToNextStep = /**
         * Tries to transition the wizard to the next step from the `currentStep`
         * @param {?=} preFinalize
         * @param {?=} postFinalize
         * @return {?}
         */
            function (preFinalize, postFinalize) {
                if (this.wizardState.hasNextStep()) {
                    this.goToStep(this.wizardState.currentStepIndex + 1, preFinalize, postFinalize);
                }
            };
        return NavigationMode;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * A [[NavigationMode]], which allows the user to navigate without any limitations,
     * as long as the current step can be exited in the given direction
     *
     * @author Marc Arndt
     */
    var /**
     * A [[NavigationMode]], which allows the user to navigate without any limitations,
     * as long as the current step can be exited in the given direction
     *
     * @author Marc Arndt
     */ FreeNavigationMode = /** @class */ (function (_super) {
        __extends(FreeNavigationMode, _super);
        /**
         * Constructor
         *
         * @param wizardState The model/state of the wizard, that is configured with this navigation mode
         */
        function FreeNavigationMode(wizardState) {
            return _super.call(this, wizardState) || this;
        }
        /**
         * Checks whether the wizard can be transitioned to the given destination step.
         * A destination wizard step can be entered if:
         * - it exists
         * - the current step can be exited in the direction of the destination step
         *
         * @param destinationIndex The index of the destination wizard step
         * @returns True if the destination wizard step can be entered, false otherwise
         */
        /**
         * Checks whether the wizard can be transitioned to the given destination step.
         * A destination wizard step can be entered if:
         * - it exists
         * - the current step can be exited in the direction of the destination step
         *
         * @param {?} destinationIndex The index of the destination wizard step
         * @return {?} True if the destination wizard step can be entered, false otherwise
         */
        FreeNavigationMode.prototype.canGoToStep = /**
         * Checks whether the wizard can be transitioned to the given destination step.
         * A destination wizard step can be entered if:
         * - it exists
         * - the current step can be exited in the direction of the destination step
         *
         * @param {?} destinationIndex The index of the destination wizard step
         * @return {?} True if the destination wizard step can be entered, false otherwise
         */
            function (destinationIndex) {
                var _this = this;
                /** @type {?} */
                var hasStep = this.wizardState.hasStep(destinationIndex);
                /** @type {?} */
                var movingDirection = this.wizardState.getMovingDirection(destinationIndex);
                /** @type {?} */
                var canExitCurrentStep = ( /**
                 * @param {?} previous
                 * @return {?}
                 */function (previous) {
                    return previous ? _this.wizardState.currentStep.canExitStep(movingDirection) : Promise.resolve(false);
                });
                /** @type {?} */
                var canEnterDestinationStep = ( /**
                 * @param {?} previous
                 * @return {?}
                 */function (previous) {
                    return previous ? _this.wizardState.getStepAtIndex(destinationIndex).canEnterStep(movingDirection) : Promise.resolve(false);
                });
                return Promise.resolve(hasStep)
                    .then(canExitCurrentStep)
                    .then(canEnterDestinationStep);
            };
        /**
         * Tries to enter the wizard step with the given destination index.
         * When entering the destination step, the following actions are done:
         * - the old current step is set as completed
         * - the old current step is set as unselected
         * - the old current step is exited
         * - the destination step is set as selected
         * - the destination step is entered
         *
         * When the destination step couldn't be entered, the following actions are done:
         * - the current step is exited and entered in the direction `MovingDirection.Stay`
         *
         * @param destinationIndex The index of the destination wizard step, which should be entered
         * @param preFinalize An event emitter, to be called before the step has been transitioned
         * @param postFinalize An event emitter, to be called after the step has been transitioned
         */
        /**
         * Tries to enter the wizard step with the given destination index.
         * When entering the destination step, the following actions are done:
         * - the old current step is set as completed
         * - the old current step is set as unselected
         * - the old current step is exited
         * - the destination step is set as selected
         * - the destination step is entered
         *
         * When the destination step couldn't be entered, the following actions are done:
         * - the current step is exited and entered in the direction `MovingDirection.Stay`
         *
         * @param {?} destinationIndex The index of the destination wizard step, which should be entered
         * @param {?=} preFinalize An event emitter, to be called before the step has been transitioned
         * @param {?=} postFinalize An event emitter, to be called after the step has been transitioned
         * @return {?}
         */
        FreeNavigationMode.prototype.goToStep = /**
         * Tries to enter the wizard step with the given destination index.
         * When entering the destination step, the following actions are done:
         * - the old current step is set as completed
         * - the old current step is set as unselected
         * - the old current step is exited
         * - the destination step is set as selected
         * - the destination step is entered
         *
         * When the destination step couldn't be entered, the following actions are done:
         * - the current step is exited and entered in the direction `MovingDirection.Stay`
         *
         * @param {?} destinationIndex The index of the destination wizard step, which should be entered
         * @param {?=} preFinalize An event emitter, to be called before the step has been transitioned
         * @param {?=} postFinalize An event emitter, to be called after the step has been transitioned
         * @return {?}
         */
            function (destinationIndex, preFinalize, postFinalize) {
                var _this = this;
                this.canGoToStep(destinationIndex).then(( /**
                 * @param {?} navigationAllowed
                 * @return {?}
                 */function (navigationAllowed) {
                    if (navigationAllowed) {
                        // the current step can be exited in the given direction
                        /** @type {?} */
                        var movingDirection = _this.wizardState.getMovingDirection(destinationIndex);
                        /* istanbul ignore if */
                        if (preFinalize) {
                            preFinalize.emit();
                        }
                        // leave current step
                        _this.wizardState.currentStep.completed = true;
                        _this.wizardState.currentStep.exit(movingDirection);
                        _this.wizardState.currentStep.selected = false;
                        _this.wizardState.currentStepIndex = destinationIndex;
                        // go to next step
                        _this.wizardState.currentStep.enter(movingDirection);
                        _this.wizardState.currentStep.selected = true;
                        /* istanbul ignore if */
                        if (postFinalize) {
                            postFinalize.emit();
                        }
                    }
                    else {
                        // if the current step can't be left, reenter the current step
                        _this.wizardState.currentStep.exit(MovingDirection.Stay);
                        _this.wizardState.currentStep.enter(MovingDirection.Stay);
                    }
                }));
            };
        /**
         * @param {?} destinationIndex
         * @return {?}
         */
        FreeNavigationMode.prototype.isNavigable = /**
         * @param {?} destinationIndex
         * @return {?}
         */
            function (destinationIndex) {
                return true;
            };
        /**
         * Resets the state of this wizard.
         * A reset transitions the wizard automatically to the first step and sets all steps as incomplete.
         * In addition the whole wizard is set as incomplete
         */
        /**
         * Resets the state of this wizard.
         * A reset transitions the wizard automatically to the first step and sets all steps as incomplete.
         * In addition the whole wizard is set as incomplete
         * @return {?}
         */
        FreeNavigationMode.prototype.reset = /**
         * Resets the state of this wizard.
         * A reset transitions the wizard automatically to the first step and sets all steps as incomplete.
         * In addition the whole wizard is set as incomplete
         * @return {?}
         */
            function () {
                // the wizard doesn't contain a step with the default step index
                if (!this.wizardState.hasStep(this.wizardState.defaultStepIndex)) {
                    throw new Error("The wizard doesn't contain a step with index " + this.wizardState.defaultStepIndex);
                }
                // reset the step internal state
                this.wizardState.wizardSteps.forEach(( /**
                 * @param {?} step
                 * @return {?}
                 */function (step) {
                    step.completed = false;
                    step.selected = false;
                }));
                // set the first step as the current step
                this.wizardState.currentStepIndex = this.wizardState.defaultStepIndex;
                this.wizardState.currentStep.selected = true;
                this.wizardState.currentStep.enter(MovingDirection.Forwards);
            };
        return FreeNavigationMode;
    }(NavigationMode));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * A [[NavigationMode]], which allows the user to navigate with some limitations.
     * The user can only navigation to a given destination step, if:
     * - the current step can be exited in the direction of the destination step
     * - a completion step can only be entered, if all "normal" wizard steps have been completed
     *
     * @author Marc Arndt
     */
    var /**
     * A [[NavigationMode]], which allows the user to navigate with some limitations.
     * The user can only navigation to a given destination step, if:
     * - the current step can be exited in the direction of the destination step
     * - a completion step can only be entered, if all "normal" wizard steps have been completed
     *
     * @author Marc Arndt
     */ SemiStrictNavigationMode = /** @class */ (function (_super) {
        __extends(SemiStrictNavigationMode, _super);
        /**
         * Constructor
         *
         * @param wizardState The model/state of the wizard, that is configured with this navigation mode
         */
        function SemiStrictNavigationMode(wizardState) {
            return _super.call(this, wizardState) || this;
        }
        /**
         * Checks whether the wizard can be transitioned to the given destination step.
         * A destination wizard step can be entered if:
         * - it exists
         * - the current step can be exited in the direction of the destination step
         * - all "normal" wizard steps have been completed, are optional or selected, or the destination step isn't a completion step
         *
         * @param destinationIndex The index of the destination wizard step
         * @returns True if the destination wizard step can be entered, false otherwise
         */
        /**
         * Checks whether the wizard can be transitioned to the given destination step.
         * A destination wizard step can be entered if:
         * - it exists
         * - the current step can be exited in the direction of the destination step
         * - all "normal" wizard steps have been completed, are optional or selected, or the destination step isn't a completion step
         *
         * @param {?} destinationIndex The index of the destination wizard step
         * @return {?} True if the destination wizard step can be entered, false otherwise
         */
        SemiStrictNavigationMode.prototype.canGoToStep = /**
         * Checks whether the wizard can be transitioned to the given destination step.
         * A destination wizard step can be entered if:
         * - it exists
         * - the current step can be exited in the direction of the destination step
         * - all "normal" wizard steps have been completed, are optional or selected, or the destination step isn't a completion step
         *
         * @param {?} destinationIndex The index of the destination wizard step
         * @return {?} True if the destination wizard step can be entered, false otherwise
         */
            function (destinationIndex) {
                var _this = this;
                /** @type {?} */
                var hasStep = this.wizardState.hasStep(destinationIndex);
                /** @type {?} */
                var movingDirection = this.wizardState.getMovingDirection(destinationIndex);
                /** @type {?} */
                var canExitCurrentStep = ( /**
                 * @param {?} previous
                 * @return {?}
                 */function (previous) {
                    return previous ? _this.wizardState.currentStep.canExitStep(movingDirection) : Promise.resolve(false);
                });
                /** @type {?} */
                var canEnterDestinationStep = ( /**
                 * @param {?} previous
                 * @return {?}
                 */function (previous) {
                    return previous ? _this.wizardState.getStepAtIndex(destinationIndex).canEnterStep(movingDirection) : Promise.resolve(false);
                });
                // provide the destination step as a lambda in case the index doesn't exist (i.e. hasStep === false)
                /** @type {?} */
                var destinationStep = ( /**
                 * @param {?} previous
                 * @return {?}
                 */function (previous) {
                    if (previous) {
                        /** @type {?} */
                        var allNormalStepsCompleted = _this.wizardState.wizardSteps
                            .filter(( /**
                     * @param {?} step
                     * @param {?} index
                     * @return {?}
                     */function (step, index) { return index < destinationIndex; }))
                            .every(( /**
                     * @param {?} step
                     * @return {?}
                     */function (step) { return step.completed || step.optional || step.selected; }));
                        return Promise.resolve(!(_this.wizardState.getStepAtIndex(destinationIndex) instanceof WizardCompletionStep) || allNormalStepsCompleted);
                    }
                    else {
                        return Promise.resolve(false);
                    }
                });
                return Promise.resolve(hasStep)
                    .then(canExitCurrentStep)
                    .then(canEnterDestinationStep)
                    .then(destinationStep);
            };
        /**
         * Tries to enter the wizard step with the given destination index.
         * When entering the destination step, the following actions are done:
         * - the old current step is set as completed
         * - the old current step is set as unselected
         * - the old current step is exited
         * - the destination step is set as selected
         * - the destination step is entered
         *
         * When the destination step couldn't be entered, the following actions are done:
         * - the current step is exited and entered in the direction `MovingDirection.Stay`
         *
         * @param destinationIndex The index of the destination wizard step, which should be entered
         * @param preFinalize An event emitter, to be called before the step has been transitioned
         * @param postFinalize An event emitter, to be called after the step has been transitioned
         */
        /**
         * Tries to enter the wizard step with the given destination index.
         * When entering the destination step, the following actions are done:
         * - the old current step is set as completed
         * - the old current step is set as unselected
         * - the old current step is exited
         * - the destination step is set as selected
         * - the destination step is entered
         *
         * When the destination step couldn't be entered, the following actions are done:
         * - the current step is exited and entered in the direction `MovingDirection.Stay`
         *
         * @param {?} destinationIndex The index of the destination wizard step, which should be entered
         * @param {?=} preFinalize An event emitter, to be called before the step has been transitioned
         * @param {?=} postFinalize An event emitter, to be called after the step has been transitioned
         * @return {?}
         */
        SemiStrictNavigationMode.prototype.goToStep = /**
         * Tries to enter the wizard step with the given destination index.
         * When entering the destination step, the following actions are done:
         * - the old current step is set as completed
         * - the old current step is set as unselected
         * - the old current step is exited
         * - the destination step is set as selected
         * - the destination step is entered
         *
         * When the destination step couldn't be entered, the following actions are done:
         * - the current step is exited and entered in the direction `MovingDirection.Stay`
         *
         * @param {?} destinationIndex The index of the destination wizard step, which should be entered
         * @param {?=} preFinalize An event emitter, to be called before the step has been transitioned
         * @param {?=} postFinalize An event emitter, to be called after the step has been transitioned
         * @return {?}
         */
            function (destinationIndex, preFinalize, postFinalize) {
                var _this = this;
                this.canGoToStep(destinationIndex).then(( /**
                 * @param {?} navigationAllowed
                 * @return {?}
                 */function (navigationAllowed) {
                    if (navigationAllowed) {
                        // the current step can be exited in the given direction
                        /** @type {?} */
                        var movingDirection = _this.wizardState.getMovingDirection(destinationIndex);
                        /* istanbul ignore if */
                        if (preFinalize) {
                            preFinalize.emit();
                        }
                        // leave current step
                        _this.wizardState.currentStep.completed = true;
                        _this.wizardState.currentStep.exit(movingDirection);
                        _this.wizardState.currentStep.selected = false;
                        _this.wizardState.currentStepIndex = destinationIndex;
                        // go to next step
                        _this.wizardState.currentStep.enter(movingDirection);
                        _this.wizardState.currentStep.selected = true;
                        /* istanbul ignore if */
                        if (postFinalize) {
                            postFinalize.emit();
                        }
                    }
                    else {
                        // if the current step can't be left, reenter the current step
                        _this.wizardState.currentStep.exit(MovingDirection.Stay);
                        _this.wizardState.currentStep.enter(MovingDirection.Stay);
                    }
                }));
            };
        /**
         * @inheritDoc
         */
        /**
         * @inheritDoc
         * @param {?} destinationIndex
         * @return {?}
         */
        SemiStrictNavigationMode.prototype.isNavigable = /**
         * @inheritDoc
         * @param {?} destinationIndex
         * @return {?}
         */
            function (destinationIndex) {
                if (this.wizardState.getStepAtIndex(destinationIndex) instanceof WizardCompletionStep) {
                    // a completion step can only be entered, if all previous steps have been completed, are optional, or selected
                    return this.wizardState.wizardSteps.filter(( /**
                     * @param {?} step
                     * @param {?} index
                     * @return {?}
                     */function (step, index) { return index < destinationIndex; }))
                        .every(( /**
                 * @param {?} step
                 * @return {?}
                 */function (step) { return step.completed || step.optional || step.selected; }));
                }
                else {
                    // a "normal" step can always be entered
                    return true;
                }
            };
        /**
         * @inheritDoc
         */
        /**
         * @inheritDoc
         * @return {?}
         */
        SemiStrictNavigationMode.prototype.reset = /**
         * @inheritDoc
         * @return {?}
         */
            function () {
                // the wizard doesn't contain a step with the default step index
                if (!this.wizardState.hasStep(this.wizardState.defaultStepIndex)) {
                    throw new Error("The wizard doesn't contain a step with index " + this.wizardState.defaultStepIndex);
                }
                // the default step is a completion step and the wizard contains more than one step
                /** @type {?} */
                var defaultCompletionStep = this.wizardState.getStepAtIndex(this.wizardState.defaultStepIndex) instanceof WizardCompletionStep &&
                    this.wizardState.wizardSteps.length !== 1;
                if (defaultCompletionStep) {
                    throw new Error("The default step index " + this.wizardState.defaultStepIndex + " references a completion step");
                }
                // reset the step internal state
                this.wizardState.wizardSteps.forEach(( /**
                 * @param {?} step
                 * @return {?}
                 */function (step) {
                    step.completed = false;
                    step.selected = false;
                }));
                // set the first step as the current step
                this.wizardState.currentStepIndex = this.wizardState.defaultStepIndex;
                this.wizardState.currentStep.selected = true;
                this.wizardState.currentStep.enter(MovingDirection.Forwards);
            };
        return SemiStrictNavigationMode;
    }(NavigationMode));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * A [[NavigationMode]], which allows the user to navigate with strict limitations.
     * The user can only navigation to a given destination step, if:
     * - the current step can be exited in the direction of the destination step
     * - all previous steps to the destination step have been completed or are optional
     *
     * @author Marc Arndt
     */
    var /**
     * A [[NavigationMode]], which allows the user to navigate with strict limitations.
     * The user can only navigation to a given destination step, if:
     * - the current step can be exited in the direction of the destination step
     * - all previous steps to the destination step have been completed or are optional
     *
     * @author Marc Arndt
     */ StrictNavigationMode = /** @class */ (function (_super) {
        __extends(StrictNavigationMode, _super);
        /**
         * Constructor
         *
         * @param wizardState The state of the wizard, that is configured with this navigation mode
         */
        function StrictNavigationMode(wizardState) {
            return _super.call(this, wizardState) || this;
        }
        /**
         * Checks whether the wizard can be transitioned to the given destination step.
         * A destination wizard step can be entered if:
         * - it exists
         * - the current step can be exited in the direction of the destination step
         * - all previous steps to the destination step have been completed or are optional
         *
         * @param destinationIndex The index of the destination wizard step
         * @returns True if the destination wizard step can be entered, false otherwise
         */
        /**
         * Checks whether the wizard can be transitioned to the given destination step.
         * A destination wizard step can be entered if:
         * - it exists
         * - the current step can be exited in the direction of the destination step
         * - all previous steps to the destination step have been completed or are optional
         *
         * @param {?} destinationIndex The index of the destination wizard step
         * @return {?} True if the destination wizard step can be entered, false otherwise
         */
        StrictNavigationMode.prototype.canGoToStep = /**
         * Checks whether the wizard can be transitioned to the given destination step.
         * A destination wizard step can be entered if:
         * - it exists
         * - the current step can be exited in the direction of the destination step
         * - all previous steps to the destination step have been completed or are optional
         *
         * @param {?} destinationIndex The index of the destination wizard step
         * @return {?} True if the destination wizard step can be entered, false otherwise
         */
            function (destinationIndex) {
                var _this = this;
                /** @type {?} */
                var hasStep = this.wizardState.hasStep(destinationIndex);
                /** @type {?} */
                var movingDirection = this.wizardState.getMovingDirection(destinationIndex);
                /** @type {?} */
                var canExitCurrentStep = ( /**
                 * @param {?} previous
                 * @return {?}
                 */function (previous) {
                    return previous ? _this.wizardState.currentStep.canExitStep(movingDirection) : Promise.resolve(false);
                });
                /** @type {?} */
                var canEnterDestinationStep = ( /**
                 * @param {?} previous
                 * @return {?}
                 */function (previous) {
                    return previous ? _this.wizardState.getStepAtIndex(destinationIndex).canEnterStep(movingDirection) : Promise.resolve(false);
                });
                /** @type {?} */
                var allPreviousStepsComplete = ( /**
                 * @param {?} previous
                 * @return {?}
                 */function (previous) {
                    if (previous) {
                        return Promise.resolve(_this.wizardState.wizardSteps
                            .filter(( /**
                     * @param {?} step
                     * @param {?} index
                     * @return {?}
                     */function (step, index) { return index < destinationIndex && index !== _this.wizardState.currentStepIndex; }))
                            .every(( /**
                     * @param {?} step
                     * @return {?}
                     */function (step) { return step.completed || step.optional; })));
                    }
                    else {
                        return Promise.resolve(false);
                    }
                });
                return Promise.resolve(hasStep)
                    .then(canExitCurrentStep)
                    .then(canEnterDestinationStep)
                    .then(allPreviousStepsComplete);
            };
        /**
         * Tries to enter the wizard step with the given destination index.
         * When entering the destination step, the following actions are done:
         * - the old current step is set as completed
         * - the old current step is set as unselected
         * - the old current step is exited
         * - all steps between the old current step and the destination step are marked as incomplete
         * - the destination step is set as selected
         * - the destination step is entered
         *
         * When the destination step couldn't be entered, the following actions are done:
         * - the current step is exited and entered in the direction `MovingDirection.Stay`
         *
         * @param destinationIndex The index of the destination wizard step, which should be entered
         * @param preFinalize An event emitter, to be called before the step has been transitioned
         * @param postFinalize An event emitter, to be called after the step has been transitioned
         */
        /**
         * Tries to enter the wizard step with the given destination index.
         * When entering the destination step, the following actions are done:
         * - the old current step is set as completed
         * - the old current step is set as unselected
         * - the old current step is exited
         * - all steps between the old current step and the destination step are marked as incomplete
         * - the destination step is set as selected
         * - the destination step is entered
         *
         * When the destination step couldn't be entered, the following actions are done:
         * - the current step is exited and entered in the direction `MovingDirection.Stay`
         *
         * @param {?} destinationIndex The index of the destination wizard step, which should be entered
         * @param {?=} preFinalize An event emitter, to be called before the step has been transitioned
         * @param {?=} postFinalize An event emitter, to be called after the step has been transitioned
         * @return {?}
         */
        StrictNavigationMode.prototype.goToStep = /**
         * Tries to enter the wizard step with the given destination index.
         * When entering the destination step, the following actions are done:
         * - the old current step is set as completed
         * - the old current step is set as unselected
         * - the old current step is exited
         * - all steps between the old current step and the destination step are marked as incomplete
         * - the destination step is set as selected
         * - the destination step is entered
         *
         * When the destination step couldn't be entered, the following actions are done:
         * - the current step is exited and entered in the direction `MovingDirection.Stay`
         *
         * @param {?} destinationIndex The index of the destination wizard step, which should be entered
         * @param {?=} preFinalize An event emitter, to be called before the step has been transitioned
         * @param {?=} postFinalize An event emitter, to be called after the step has been transitioned
         * @return {?}
         */
            function (destinationIndex, preFinalize, postFinalize) {
                var _this = this;
                this.canGoToStep(destinationIndex).then(( /**
                 * @param {?} navigationAllowed
                 * @return {?}
                 */function (navigationAllowed) {
                    if (navigationAllowed) {
                        /** @type {?} */
                        var movingDirection = _this.wizardState.getMovingDirection(destinationIndex);
                        /* istanbul ignore if */
                        if (preFinalize) {
                            preFinalize.emit();
                        }
                        // leave current step
                        _this.wizardState.currentStep.completed = true;
                        _this.wizardState.currentStep.exit(movingDirection);
                        _this.wizardState.currentStep.selected = false;
                        // set all steps after the destination step to incomplete
                        _this.wizardState.wizardSteps
                            .filter(( /**
                     * @param {?} step
                     * @param {?} index
                     * @return {?}
                     */function (step, index) { return _this.wizardState.currentStepIndex > destinationIndex && index > destinationIndex; }))
                            .forEach(( /**
                     * @param {?} step
                     * @return {?}
                     */function (step) { return step.completed = false; }));
                        _this.wizardState.currentStepIndex = destinationIndex;
                        // go to next step
                        _this.wizardState.currentStep.enter(movingDirection);
                        _this.wizardState.currentStep.selected = true;
                        /* istanbul ignore if */
                        if (postFinalize) {
                            postFinalize.emit();
                        }
                    }
                    else {
                        // if the current step can't be left, reenter the current step
                        _this.wizardState.currentStep.exit(MovingDirection.Stay);
                        _this.wizardState.currentStep.enter(MovingDirection.Stay);
                    }
                }));
            };
        /**
         * @param {?} destinationIndex
         * @return {?}
         */
        StrictNavigationMode.prototype.isNavigable = /**
         * @param {?} destinationIndex
         * @return {?}
         */
            function (destinationIndex) {
                // a wizard step can be navigated to through the navigation bar, iff it's located before the current wizard step
                return destinationIndex < this.wizardState.currentStepIndex;
            };
        /**
         * Resets the state of this wizard.
         * A reset transitions the wizard automatically to the first step and sets all steps as incomplete.
         * In addition the whole wizard is set as incomplete
         */
        /**
         * Resets the state of this wizard.
         * A reset transitions the wizard automatically to the first step and sets all steps as incomplete.
         * In addition the whole wizard is set as incomplete
         * @return {?}
         */
        StrictNavigationMode.prototype.reset = /**
         * Resets the state of this wizard.
         * A reset transitions the wizard automatically to the first step and sets all steps as incomplete.
         * In addition the whole wizard is set as incomplete
         * @return {?}
         */
            function () {
                var _this = this;
                // the wizard doesn't contain a step with the default step index
                if (!this.wizardState.hasStep(this.wizardState.defaultStepIndex)) {
                    throw new Error("The wizard doesn't contain a step with index " + this.wizardState.defaultStepIndex);
                }
                // at least one step is before the default step, that is not optional
                /** @type {?} */
                var illegalDefaultStep = this.wizardState.wizardSteps
                    .filter(( /**
             * @param {?} step
             * @param {?} index
             * @return {?}
             */function (step, index) { return index < _this.wizardState.defaultStepIndex; }))
                    .some(( /**
             * @param {?} step
             * @return {?}
             */function (step) { return !step.optional; }));
                if (illegalDefaultStep) {
                    throw new Error("The default step index " + this.wizardState.defaultStepIndex + " is located after a non optional step");
                }
                // reset the step internal state
                this.wizardState.wizardSteps.forEach(( /**
                 * @param {?} step
                 * @return {?}
                 */function (step) {
                    step.completed = false;
                    step.selected = false;
                }));
                // set the first step as the current step
                this.wizardState.currentStepIndex = this.wizardState.defaultStepIndex;
                this.wizardState.currentStep.selected = true;
                this.wizardState.currentStep.enter(MovingDirection.Forwards);
            };
        return StrictNavigationMode;
    }(NavigationMode));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * A factory method used to create [[NavigationMode]] instances
     *
     * @param {?} navigationMode The name of the to be used navigation mode
     * @param {?} wizardState The wizard state of the wizard
     * @return {?} The created [[NavigationMode]]
     */
    function navigationModeFactory(navigationMode, wizardState) {
        switch (navigationMode) {
            case 'free':
                return new FreeNavigationMode(wizardState);
            case 'semi-strict':
                return new SemiStrictNavigationMode(wizardState);
            case 'strict':
            default:
                return new StrictNavigationMode(wizardState);
        }
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The internal model/state of a wizard.
     * This model contains:
     * - an array with all steps the wizard contains
     * - the index of the step the wizard currently resides inside
     * - information about the completeness of the wizard
     * - some additional helper methods
     *
     * @author Marc Arndt
     */
    var WizardState = /** @class */ (function () {
        /**
         * Constructor
         */
        function WizardState() {
            /**
             * An array representation of all wizard steps belonging to this model
             */
            this.wizardSteps = [];
            /**
             * The index of the currently visible and selected step inside the wizardSteps QueryList.
             * If this wizard contains no steps, currentStepIndex is -1
             */
            this.currentStepIndex = -1;
            /**
             * The initial step index, as taken from the [[WizardComponent]]
             */
            this._defaultStepIndex = 0;
        }
        Object.defineProperty(WizardState.prototype, "defaultStepIndex", {
            /**
             * The initial step index.
             * This value can be either:
             * - the index of a wizard step with a `selected` directive, or
             * - the default step index, set in the [[WizardComponent]]
             */
            get: /**
             * The initial step index.
             * This value can be either:
             * - the index of a wizard step with a `selected` directive, or
             * - the default step index, set in the [[WizardComponent]]
             * @return {?}
             */ function () {
                /** @type {?} */
                var foundDefaultStep = this.wizardSteps.find(( /**
                 * @param {?} step
                 * @return {?}
                 */function (step) { return step.defaultSelected; }));
                if (foundDefaultStep) {
                    return this.getIndexOfStep(foundDefaultStep);
                }
                else {
                    return this._defaultStepIndex;
                }
            },
            /**
             * Sets the initial default step.
             * Beware: This initial default is only used if no wizard step has been enhanced with the `selected` directive
             *
             * @param defaultStepIndex The new default wizard step index
             */
            set: /**
             * Sets the initial default step.
             * Beware: This initial default is only used if no wizard step has been enhanced with the `selected` directive
             *
             * @param {?} defaultStepIndex The new default wizard step index
             * @return {?}
             */ function (defaultStepIndex) {
                this._defaultStepIndex = defaultStepIndex;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(WizardState.prototype, "currentStep", {
            /**
             * The WizardStep object belonging to the currently visible and selected step.
             * The currentStep is always the currently selected wizard step.
             * The currentStep can be either completed, if it was visited earlier,
             * or not completed, if it is visited for the first time or its state is currently out of date.
             *
             * If this wizard contains no steps, currentStep is null
             */
            get: /**
             * The WizardStep object belonging to the currently visible and selected step.
             * The currentStep is always the currently selected wizard step.
             * The currentStep can be either completed, if it was visited earlier,
             * or not completed, if it is visited for the first time or its state is currently out of date.
             *
             * If this wizard contains no steps, currentStep is null
             * @return {?}
             */ function () {
                if (this.hasStep(this.currentStepIndex)) {
                    return this.wizardSteps[this.currentStepIndex];
                }
                else {
                    return null;
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(WizardState.prototype, "completed", {
            /**
             * The completeness of the wizard.
             * If the wizard has been completed, i.e. all steps are either completed or optional, this value is true, otherwise it is false
             */
            get: /**
             * The completeness of the wizard.
             * If the wizard has been completed, i.e. all steps are either completed or optional, this value is true, otherwise it is false
             * @return {?}
             */ function () {
                return this.wizardSteps.every(( /**
                 * @param {?} step
                 * @return {?}
                 */function (step) { return step.completed || step.optional; }));
            },
            enumerable: true,
            configurable: true
        });
        /**
         * Updates the navigation mode to the navigation mode with the given name
         *
         * @param updatedNavigationMode The name of the new navigation mode
         */
        /**
         * Updates the navigation mode to the navigation mode with the given name
         *
         * @param {?} updatedNavigationMode The name of the new navigation mode
         * @return {?}
         */
        WizardState.prototype.updateNavigationMode = /**
         * Updates the navigation mode to the navigation mode with the given name
         *
         * @param {?} updatedNavigationMode The name of the new navigation mode
         * @return {?}
         */
            function (updatedNavigationMode) {
                this.navigationMode = navigationModeFactory(updatedNavigationMode, this);
            };
        /**
         * Updates the wizard steps to the new array
         *
         * @param updatedWizardSteps The updated wizard steps
         */
        /**
         * Updates the wizard steps to the new array
         *
         * @param {?} updatedWizardSteps The updated wizard steps
         * @return {?}
         */
        WizardState.prototype.updateWizardSteps = /**
         * Updates the wizard steps to the new array
         *
         * @param {?} updatedWizardSteps The updated wizard steps
         * @return {?}
         */
            function (updatedWizardSteps) {
                // the wizard is currently not in the initialization phase
                if (this.wizardSteps.length > 0 && this.currentStepIndex > -1) {
                    this.currentStepIndex = updatedWizardSteps.indexOf(this.wizardSteps[this.currentStepIndex]);
                }
                this.wizardSteps = updatedWizardSteps;
            };
        /**
         * Checks if a given index `stepIndex` is inside the range of possible wizard steps inside this wizard
         *
         * @param stepIndex The to be checked index of a step inside this wizard
         * @returns True if the given `stepIndex` is contained inside this wizard, false otherwise
         */
        /**
         * Checks if a given index `stepIndex` is inside the range of possible wizard steps inside this wizard
         *
         * @param {?} stepIndex The to be checked index of a step inside this wizard
         * @return {?} True if the given `stepIndex` is contained inside this wizard, false otherwise
         */
        WizardState.prototype.hasStep = /**
         * Checks if a given index `stepIndex` is inside the range of possible wizard steps inside this wizard
         *
         * @param {?} stepIndex The to be checked index of a step inside this wizard
         * @return {?} True if the given `stepIndex` is contained inside this wizard, false otherwise
         */
            function (stepIndex) {
                return this.wizardSteps.length > 0 && 0 <= stepIndex && stepIndex < this.wizardSteps.length;
            };
        /**
         * Checks if this wizard has a previous step, compared to the current step
         *
         * @returns True if this wizard has a previous step before the current step
         */
        /**
         * Checks if this wizard has a previous step, compared to the current step
         *
         * @return {?} True if this wizard has a previous step before the current step
         */
        WizardState.prototype.hasPreviousStep = /**
         * Checks if this wizard has a previous step, compared to the current step
         *
         * @return {?} True if this wizard has a previous step before the current step
         */
            function () {
                return this.hasStep(this.currentStepIndex - 1);
            };
        /**
         * Checks if this wizard has a next step, compared to the current step
         *
         * @returns True if this wizard has a next step after the current step
         */
        /**
         * Checks if this wizard has a next step, compared to the current step
         *
         * @return {?} True if this wizard has a next step after the current step
         */
        WizardState.prototype.hasNextStep = /**
         * Checks if this wizard has a next step, compared to the current step
         *
         * @return {?} True if this wizard has a next step after the current step
         */
            function () {
                return this.hasStep(this.currentStepIndex + 1);
            };
        /**
         * Checks if this wizard is currently inside its last step
         *
         * @returns True if the wizard is currently inside its last step
         */
        /**
         * Checks if this wizard is currently inside its last step
         *
         * @return {?} True if the wizard is currently inside its last step
         */
        WizardState.prototype.isLastStep = /**
         * Checks if this wizard is currently inside its last step
         *
         * @return {?} True if the wizard is currently inside its last step
         */
            function () {
                return this.wizardSteps.length > 0 && this.currentStepIndex === this.wizardSteps.length - 1;
            };
        /**
         * Finds the [[WizardStep]] at the given index `stepIndex`.
         * If no [[WizardStep]] exists at the given index an Error is thrown
         *
         * @param stepIndex The given index
         * @returns The found [[WizardStep]] at the given index `stepIndex`
         * @throws An `Error` is thrown, if the given index `stepIndex` doesn't exist
         */
        /**
         * Finds the [[WizardStep]] at the given index `stepIndex`.
         * If no [[WizardStep]] exists at the given index an Error is thrown
         *
         * @throws An `Error` is thrown, if the given index `stepIndex` doesn't exist
         * @param {?} stepIndex The given index
         * @return {?} The found [[WizardStep]] at the given index `stepIndex`
         */
        WizardState.prototype.getStepAtIndex = /**
         * Finds the [[WizardStep]] at the given index `stepIndex`.
         * If no [[WizardStep]] exists at the given index an Error is thrown
         *
         * @throws An `Error` is thrown, if the given index `stepIndex` doesn't exist
         * @param {?} stepIndex The given index
         * @return {?} The found [[WizardStep]] at the given index `stepIndex`
         */
            function (stepIndex) {
                if (!this.hasStep(stepIndex)) {
                    throw new Error("Expected a known step, but got stepIndex: " + stepIndex + ".");
                }
                return this.wizardSteps[stepIndex];
            };
        /**
         * Finds the index of the step with the given `stepId`.
         * If no step with the given `stepId` exists, `-1` is returned
         *
         * @param stepId The given step id
         * @returns The found index of a step with the given step id, or `-1` if no step with the given id is included in the wizard
         */
        /**
         * Finds the index of the step with the given `stepId`.
         * If no step with the given `stepId` exists, `-1` is returned
         *
         * @param {?} stepId The given step id
         * @return {?} The found index of a step with the given step id, or `-1` if no step with the given id is included in the wizard
         */
        WizardState.prototype.getIndexOfStepWithId = /**
         * Finds the index of the step with the given `stepId`.
         * If no step with the given `stepId` exists, `-1` is returned
         *
         * @param {?} stepId The given step id
         * @return {?} The found index of a step with the given step id, or `-1` if no step with the given id is included in the wizard
         */
            function (stepId) {
                return this.wizardSteps.findIndex(( /**
                 * @param {?} step
                 * @return {?}
                 */function (step) { return step.stepId === stepId; }));
            };
        /**
         * Finds the index of the given [[WizardStep]] `step`.
         * If the given [[WizardStep]] is not contained inside this wizard, `-1` is returned
         *
         * @param step The given [[WizardStep]]
         * @returns The found index of `step` or `-1` if the step is not included in the wizard
         */
        /**
         * Finds the index of the given [[WizardStep]] `step`.
         * If the given [[WizardStep]] is not contained inside this wizard, `-1` is returned
         *
         * @param {?} step The given [[WizardStep]]
         * @return {?} The found index of `step` or `-1` if the step is not included in the wizard
         */
        WizardState.prototype.getIndexOfStep = /**
         * Finds the index of the given [[WizardStep]] `step`.
         * If the given [[WizardStep]] is not contained inside this wizard, `-1` is returned
         *
         * @param {?} step The given [[WizardStep]]
         * @return {?} The found index of `step` or `-1` if the step is not included in the wizard
         */
            function (step) {
                return this.wizardSteps.indexOf(step);
            };
        /**
         * Calculates the correct [[MovingDirection]] value for a given `destinationStep` compared to the `currentStepIndex`.
         *
         * @param destinationStep The given destination step
         * @returns The calculated [[MovingDirection]]
         */
        /**
         * Calculates the correct [[MovingDirection]] value for a given `destinationStep` compared to the `currentStepIndex`.
         *
         * @param {?} destinationStep The given destination step
         * @return {?} The calculated [[MovingDirection]]
         */
        WizardState.prototype.getMovingDirection = /**
         * Calculates the correct [[MovingDirection]] value for a given `destinationStep` compared to the `currentStepIndex`.
         *
         * @param {?} destinationStep The given destination step
         * @return {?} The calculated [[MovingDirection]]
         */
            function (destinationStep) {
                /** @type {?} */
                var movingDirection;
                if (destinationStep > this.currentStepIndex) {
                    movingDirection = MovingDirection.Forwards;
                }
                else if (destinationStep < this.currentStepIndex) {
                    movingDirection = MovingDirection.Backwards;
                }
                else {
                    movingDirection = MovingDirection.Stay;
                }
                return movingDirection;
            };
        WizardState.decorators = [
            { type: core.Injectable }
        ];
        /** @nocollapse */
        WizardState.ctorParameters = function () { return []; };
        return WizardState;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `aw-wizard-navigation-bar` component contains the navigation bar inside a [[WizardComponent]].
     * To correctly display the navigation bar, it's required to set the right css classes for the navigation bar,
     * otherwise it will look like a normal `ul` component.
     *
     * ### Syntax
     *
     * ```html
     * <aw-wizard-navigation-bar></aw-wizard-navigation-bar>
     * ```
     *
     * @author Marc Arndt
     */
    var WizardNavigationBarComponent = /** @class */ (function () {
        /**
         * Constructor
         *
         * @param wizardState The state the wizard currently resides in
         */
        function WizardNavigationBarComponent(wizardState) {
            this.wizardState = wizardState;
            /**
             * The direction in which the wizard steps should be shown in the navigation bar.
             * This value can be either `left-to-right` or `right-to-left`
             */
            this.direction = 'left-to-right';
        }
        Object.defineProperty(WizardNavigationBarComponent.prototype, "navigationMode", {
            /**
             * The navigation mode
             */
            get: /**
             * The navigation mode
             * @return {?}
             */ function () {
                return this.wizardState.navigationMode;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(WizardNavigationBarComponent.prototype, "wizardSteps", {
            /**
             * Returns all [[WizardStep]]s contained in the wizard
             *
             * @returns An array containing all [[WizardStep]]s
             */
            get: /**
             * Returns all [[WizardStep]]s contained in the wizard
             *
             * @return {?} An array containing all [[WizardStep]]s
             */ function () {
                switch (this.direction) {
                    case 'right-to-left':
                        return this.wizardState.wizardSteps.slice().reverse();
                    case 'left-to-right':
                    default:
                        return this.wizardState.wizardSteps;
                }
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(WizardNavigationBarComponent.prototype, "numberOfWizardSteps", {
            /**
             * Returns the number of wizard steps, that need to be displaced in the navigation bar
             *
             * @returns The number of wizard steps to be displayed
             */
            get: /**
             * Returns the number of wizard steps, that need to be displaced in the navigation bar
             *
             * @return {?} The number of wizard steps to be displayed
             */ function () {
                return this.wizardState.wizardSteps.length;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * Checks, whether a [[WizardStep]] can be marked as `current` in the navigation bar
         *
         * @param wizardStep The wizard step to be checked
         * @returns True if the step can be marked as current
         */
        /**
         * Checks, whether a [[WizardStep]] can be marked as `current` in the navigation bar
         *
         * @param {?} wizardStep The wizard step to be checked
         * @return {?} True if the step can be marked as current
         */
        WizardNavigationBarComponent.prototype.isCurrent = /**
         * Checks, whether a [[WizardStep]] can be marked as `current` in the navigation bar
         *
         * @param {?} wizardStep The wizard step to be checked
         * @return {?} True if the step can be marked as current
         */
            function (wizardStep) {
                return wizardStep.selected && !wizardStep.completed && !this.wizardState.completed;
            };
        /**
         * Checks, whether a [[WizardStep]] can be marked as `done` in the navigation bar
         *
         * @param wizardStep The wizard step to be checked
         * @returns True if the step can be marked as done
         */
        /**
         * Checks, whether a [[WizardStep]] can be marked as `done` in the navigation bar
         *
         * @param {?} wizardStep The wizard step to be checked
         * @return {?} True if the step can be marked as done
         */
        WizardNavigationBarComponent.prototype.isDone = /**
         * Checks, whether a [[WizardStep]] can be marked as `done` in the navigation bar
         *
         * @param {?} wizardStep The wizard step to be checked
         * @return {?} True if the step can be marked as done
         */
            function (wizardStep) {
                return (wizardStep.completed && !wizardStep.selected) || this.wizardState.completed;
            };
        /**
         * Checks, whether a [[WizardStep]] can be marked as `default` in the navigation bar
         *
         * @param wizardStep The wizard step to be checked
         * @returns True if the step can be marked as default
         */
        /**
         * Checks, whether a [[WizardStep]] can be marked as `default` in the navigation bar
         *
         * @param {?} wizardStep The wizard step to be checked
         * @return {?} True if the step can be marked as default
         */
        WizardNavigationBarComponent.prototype.isDefault = /**
         * Checks, whether a [[WizardStep]] can be marked as `default` in the navigation bar
         *
         * @param {?} wizardStep The wizard step to be checked
         * @return {?} True if the step can be marked as default
         */
            function (wizardStep) {
                return !wizardStep.optional && !wizardStep.completed && !wizardStep.selected && !this.wizardState.completed;
            };
        /**
         * Checks, whether a [[WizardStep]] can be marked as `editing` in the navigation bar
         *
         * @param wizardStep The wizard step to be checked
         * @returns True if the step can be marked as editing
         */
        /**
         * Checks, whether a [[WizardStep]] can be marked as `editing` in the navigation bar
         *
         * @param {?} wizardStep The wizard step to be checked
         * @return {?} True if the step can be marked as editing
         */
        WizardNavigationBarComponent.prototype.isEditing = /**
         * Checks, whether a [[WizardStep]] can be marked as `editing` in the navigation bar
         *
         * @param {?} wizardStep The wizard step to be checked
         * @return {?} True if the step can be marked as editing
         */
            function (wizardStep) {
                return wizardStep.selected && wizardStep.completed && !this.wizardState.completed;
            };
        /**
         * Checks, whether a [[WizardStep]] can be marked as `optional` in the navigation bar
         *
         * @param wizardStep The wizard step to be checked
         * @returns True if the step can be marked as optional
         */
        /**
         * Checks, whether a [[WizardStep]] can be marked as `optional` in the navigation bar
         *
         * @param {?} wizardStep The wizard step to be checked
         * @return {?} True if the step can be marked as optional
         */
        WizardNavigationBarComponent.prototype.isOptional = /**
         * Checks, whether a [[WizardStep]] can be marked as `optional` in the navigation bar
         *
         * @param {?} wizardStep The wizard step to be checked
         * @return {?} True if the step can be marked as optional
         */
            function (wizardStep) {
                return wizardStep.optional && !wizardStep.completed && !wizardStep.selected && !this.wizardState.completed;
            };
        /**
         * Checks, whether a [[WizardStep]] can be marked as `navigable` in the navigation bar.
         * A wizard step can be navigated to if:
         * - the step is currently not selected
         * - the navigation bar isn't disabled
         * - the navigation mode allows navigation to the step
         *
         * @param wizardStep The wizard step to be checked
         * @returns True if the step can be marked as navigable
         */
        /**
         * Checks, whether a [[WizardStep]] can be marked as `navigable` in the navigation bar.
         * A wizard step can be navigated to if:
         * - the step is currently not selected
         * - the navigation bar isn't disabled
         * - the navigation mode allows navigation to the step
         *
         * @param {?} wizardStep The wizard step to be checked
         * @return {?} True if the step can be marked as navigable
         */
        WizardNavigationBarComponent.prototype.isNavigable = /**
         * Checks, whether a [[WizardStep]] can be marked as `navigable` in the navigation bar.
         * A wizard step can be navigated to if:
         * - the step is currently not selected
         * - the navigation bar isn't disabled
         * - the navigation mode allows navigation to the step
         *
         * @param {?} wizardStep The wizard step to be checked
         * @return {?} True if the step can be marked as navigable
         */
            function (wizardStep) {
                return !wizardStep.selected && !this.wizardState.disableNavigationBar &&
                    this.navigationMode.isNavigable(this.wizardState.getIndexOfStep(wizardStep));
            };
        WizardNavigationBarComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'aw-wizard-navigation-bar',
                        template: "<ul class=\"steps-indicator steps-{{numberOfWizardSteps}}\">\n  <li [attr.id]=\"step.stepId\" *ngFor=\"let step of wizardSteps\"\n      [ngClass]=\"{\n        default: isDefault(step),\n        current: isCurrent(step),\n        done: isDone(step),\n        editing: isEditing(step),\n        optional: isOptional(step),\n        navigable: isNavigable(step)\n  }\">\n    <a [awGoToStep]=\"step\">\n      <div class=\"label\">\n        <ng-container *ngIf=\"step.stepTitleTemplate\" [ngTemplateOutlet]=\"step.stepTitleTemplate.templateRef\"></ng-container>\n        <ng-container *ngIf=\"!step.stepTitleTemplate\">{{step.stepTitle}}</ng-container>\n      </div>\n      <div class=\"step-indicator\" [ngStyle]=\"{ 'font-family': step.stepSymbolTemplate ? '' : step.navigationSymbol.fontFamily }\">\n        <ng-container *ngIf=\"step.stepSymbolTemplate\" [ngTemplateOutlet]=\"step.stepSymbolTemplate.templateRef\"></ng-container>\n        <ng-container *ngIf=\"!step.stepSymbolTemplate\">{{step.navigationSymbol.symbol}}</ng-container>\n      </div>\n    </a>\n  </li>\n</ul>\n",
                        encapsulation: core.ViewEncapsulation.None,
                        styles: ["aw-wizard-navigation-bar.horizontal.small ul.steps-indicator{padding:24px 0 10px}aw-wizard-navigation-bar.horizontal.small ul.steps-indicator li:not(:last-child):before{background-color:#e6e6e6;content:'';position:absolute;height:1px;width:calc(100% - 14px);top:-7px;left:calc(50% + 14px / 2)}aw-wizard-navigation-bar.horizontal.small ul.steps-indicator li .step-indicator{position:absolute;top:-14px;left:calc(50% - 14px / 2);width:14px;height:14px;text-align:center;vertical-align:middle;line-height:14px;transition:.25s;border-radius:100%;background-color:#e6e6e6}aw-wizard-navigation-bar.horizontal.small ul.steps-indicator li.current .step-indicator{background-color:grey}aw-wizard-navigation-bar.horizontal.small ul.steps-indicator li.done .step-indicator{background-color:#393}aw-wizard-navigation-bar.horizontal.small ul.steps-indicator li.optional .step-indicator{background-color:#38ef38}aw-wizard-navigation-bar.horizontal.small ul.steps-indicator li.editing .step-indicator{background-color:red}aw-wizard-navigation-bar.horizontal.small ul.steps-indicator li a:hover .step-indicator{background-color:#d9d9d9}aw-wizard-navigation-bar.horizontal.small ul.steps-indicator li.current a:hover .step-indicator{background-color:#737373}aw-wizard-navigation-bar.horizontal.small ul.steps-indicator li.done a:hover .step-indicator{background-color:#2d862d}aw-wizard-navigation-bar.horizontal.small ul.steps-indicator li.optional a:hover .step-indicator{background-color:#20ed20}aw-wizard-navigation-bar.horizontal.small ul.steps-indicator li.editing a:hover .step-indicator{background-color:#e60000}aw-wizard-navigation-bar.horizontal.small ul.steps-indicator li.default a:hover .step-indicator{background-color:grey}aw-wizard-navigation-bar.horizontal.large-filled ul.steps-indicator{padding:60px 0 10px}aw-wizard-navigation-bar.horizontal.large-filled ul.steps-indicator li:not(:last-child):before{background-color:#e6e6e6;content:'';position:absolute;height:1px;width:calc(100% - 50px);top:-25px;left:calc(50% + 50px / 2)}aw-wizard-navigation-bar.horizontal.large-filled ul.steps-indicator li .step-indicator{position:absolute;top:-50px;left:calc(50% - 50px / 2);width:50px;height:50px;text-align:center;vertical-align:middle;line-height:50px;transition:.25s;border-radius:100%;background-color:#e6e6e6}aw-wizard-navigation-bar.horizontal.large-filled ul.steps-indicator li.current .step-indicator{background-color:grey}aw-wizard-navigation-bar.horizontal.large-filled ul.steps-indicator li.done .step-indicator{background-color:#393}aw-wizard-navigation-bar.horizontal.large-filled ul.steps-indicator li.optional .step-indicator{background-color:#38ef38}aw-wizard-navigation-bar.horizontal.large-filled ul.steps-indicator li.editing .step-indicator{background-color:red}aw-wizard-navigation-bar.horizontal.large-filled ul.steps-indicator li a:hover .step-indicator{background-color:#d9d9d9}aw-wizard-navigation-bar.horizontal.large-filled ul.steps-indicator li.current a:hover .step-indicator{background-color:#737373}aw-wizard-navigation-bar.horizontal.large-filled ul.steps-indicator li.done a:hover .step-indicator{background-color:#2d862d}aw-wizard-navigation-bar.horizontal.large-filled ul.steps-indicator li.optional a:hover .step-indicator{background-color:#20ed20}aw-wizard-navigation-bar.horizontal.large-filled ul.steps-indicator li.editing a:hover .step-indicator{background-color:#e60000}aw-wizard-navigation-bar.horizontal.large-filled ul.steps-indicator li.default a:hover .step-indicator{background-color:grey}aw-wizard-navigation-bar.horizontal.large-empty ul.steps-indicator{padding:60px 0 10px}aw-wizard-navigation-bar.horizontal.large-empty ul.steps-indicator li:not(:last-child):before{background-color:#e6e6e6;content:'';position:absolute;height:1px;width:calc(100% - 50px);top:-25px;left:calc(50% + 50px / 2)}aw-wizard-navigation-bar.horizontal.large-empty ul.steps-indicator li .step-indicator{position:absolute;top:-50px;left:calc(50% - 50px / 2);width:50px;height:50px;text-align:center;vertical-align:middle;line-height:46px;transition:.25s;border-radius:100%;border:2px solid #e6e6e6}aw-wizard-navigation-bar.horizontal.large-empty ul.steps-indicator li.current .step-indicator{border:2px solid grey}aw-wizard-navigation-bar.horizontal.large-empty ul.steps-indicator li.done .step-indicator{border:2px solid #393}aw-wizard-navigation-bar.horizontal.large-empty ul.steps-indicator li.optional .step-indicator{border:2px solid #38ef38}aw-wizard-navigation-bar.horizontal.large-empty ul.steps-indicator li.editing .step-indicator{border:2px solid red}aw-wizard-navigation-bar.horizontal.large-empty ul.steps-indicator li a:hover .step-indicator{border-color:#cdcdcd}aw-wizard-navigation-bar.horizontal.large-empty ul.steps-indicator li.current a:hover .step-indicator{border-color:#676767}aw-wizard-navigation-bar.horizontal.large-empty ul.steps-indicator li.done a:hover .step-indicator{border-color:#267326}aw-wizard-navigation-bar.horizontal.large-empty ul.steps-indicator li.optional a:hover .step-indicator{border-color:#12e212}aw-wizard-navigation-bar.horizontal.large-empty ul.steps-indicator li.editing a:hover .step-indicator{border-color:#c00}aw-wizard-navigation-bar.horizontal.large-empty ul.steps-indicator li.default a:hover .step-indicator{border-color:grey}aw-wizard-navigation-bar.horizontal.large-filled-symbols ul.steps-indicator{padding:60px 0 10px}aw-wizard-navigation-bar.horizontal.large-filled-symbols ul.steps-indicator li:not(:last-child):before{background-color:#e6e6e6;content:'';position:absolute;height:1px;width:calc(100% - 50px);top:-25px;left:calc(50% + 50px / 2)}aw-wizard-navigation-bar.horizontal.large-filled-symbols ul.steps-indicator li .step-indicator{position:absolute;top:-50px;left:calc(50% - 50px / 2);width:50px;height:50px;text-align:center;vertical-align:middle;line-height:50px;transition:.25s;border-radius:100%;background-color:#e6e6e6;color:#000}aw-wizard-navigation-bar.horizontal.large-filled-symbols ul.steps-indicator li.current .step-indicator{background-color:grey;color:#000}aw-wizard-navigation-bar.horizontal.large-filled-symbols ul.steps-indicator li.done .step-indicator{background-color:#393;color:#000}aw-wizard-navigation-bar.horizontal.large-filled-symbols ul.steps-indicator li.optional .step-indicator{background-color:#38ef38;color:#000}aw-wizard-navigation-bar.horizontal.large-filled-symbols ul.steps-indicator li.editing .step-indicator{background-color:red;color:#000}aw-wizard-navigation-bar.horizontal.large-filled-symbols ul.steps-indicator li a:hover .step-indicator{background-color:#d9d9d9}aw-wizard-navigation-bar.horizontal.large-filled-symbols ul.steps-indicator li.current a:hover .step-indicator{background-color:#737373}aw-wizard-navigation-bar.horizontal.large-filled-symbols ul.steps-indicator li.done a:hover .step-indicator{background-color:#2d862d}aw-wizard-navigation-bar.horizontal.large-filled-symbols ul.steps-indicator li.optional a:hover .step-indicator{background-color:#20ed20}aw-wizard-navigation-bar.horizontal.large-filled-symbols ul.steps-indicator li.editing a:hover .step-indicator{background-color:#e60000}aw-wizard-navigation-bar.horizontal.large-filled-symbols ul.steps-indicator li.default a:hover .step-indicator{background-color:grey}aw-wizard-navigation-bar.horizontal.large-empty-symbols ul.steps-indicator{padding:60px 0 10px}aw-wizard-navigation-bar.horizontal.large-empty-symbols ul.steps-indicator li:not(:last-child):before{background-color:#e6e6e6;content:'';position:absolute;height:1px;width:calc(100% - 50px);top:-25px;left:calc(50% + 50px / 2)}aw-wizard-navigation-bar.horizontal.large-empty-symbols ul.steps-indicator li .step-indicator{position:absolute;top:-50px;left:calc(50% - 50px / 2);width:50px;height:50px;text-align:center;vertical-align:middle;line-height:46px;transition:.25s;border-radius:100%;color:#e6e6e6;border:2px solid #e6e6e6}aw-wizard-navigation-bar.horizontal.large-empty-symbols ul.steps-indicator li.current .step-indicator{color:grey;border:2px solid grey}aw-wizard-navigation-bar.horizontal.large-empty-symbols ul.steps-indicator li.done .step-indicator{color:#393;border:2px solid #393}aw-wizard-navigation-bar.horizontal.large-empty-symbols ul.steps-indicator li.optional .step-indicator{color:#38ef38;border:2px solid #38ef38}aw-wizard-navigation-bar.horizontal.large-empty-symbols ul.steps-indicator li.editing .step-indicator{color:red;border:2px solid red}aw-wizard-navigation-bar.horizontal.large-empty-symbols ul.steps-indicator li a:hover .step-indicator{border-color:#cdcdcd;color:#cdcdcd}aw-wizard-navigation-bar.horizontal.large-empty-symbols ul.steps-indicator li.current a:hover .step-indicator{border-color:#676767;color:#676767}aw-wizard-navigation-bar.horizontal.large-empty-symbols ul.steps-indicator li.done a:hover .step-indicator{border-color:#267326;color:#267326}aw-wizard-navigation-bar.horizontal.large-empty-symbols ul.steps-indicator li.optional a:hover .step-indicator{border-color:#12e212;color:#12e212}aw-wizard-navigation-bar.horizontal.large-empty-symbols ul.steps-indicator li.editing a:hover .step-indicator{border-color:#c00;color:#c00}aw-wizard-navigation-bar.horizontal.large-empty-symbols ul.steps-indicator li.default a:hover .step-indicator{border-color:grey;color:grey}aw-wizard-navigation-bar.horizontal ul.steps-indicator{display:flex;flex-direction:row;justify-content:center;margin:0;width:100%;list-style:none}aw-wizard-navigation-bar.horizontal ul.steps-indicator.steps-2:before{left:25%;right:25%}aw-wizard-navigation-bar.horizontal ul.steps-indicator.steps-2 li{width:50%}aw-wizard-navigation-bar.horizontal ul.steps-indicator.steps-3:before{left:16.66666667%;right:16.66666667%}aw-wizard-navigation-bar.horizontal ul.steps-indicator.steps-3 li{width:33.33333333%}aw-wizard-navigation-bar.horizontal ul.steps-indicator.steps-4:before{left:12.5%;right:12.5%}aw-wizard-navigation-bar.horizontal ul.steps-indicator.steps-4 li{width:25%}aw-wizard-navigation-bar.horizontal ul.steps-indicator.steps-5:before{left:10%;right:10%}aw-wizard-navigation-bar.horizontal ul.steps-indicator.steps-5 li{width:20%}aw-wizard-navigation-bar.horizontal ul.steps-indicator.steps-6:before{left:8.33333333%;right:8.33333333%}aw-wizard-navigation-bar.horizontal ul.steps-indicator.steps-6 li{width:16.66666667%}aw-wizard-navigation-bar.horizontal ul.steps-indicator.steps-7:before{left:7.14285714%;right:7.14285714%}aw-wizard-navigation-bar.horizontal ul.steps-indicator.steps-7 li{width:14.28571429%}aw-wizard-navigation-bar.horizontal ul.steps-indicator.steps-8:before{left:6.25%;right:6.25%}aw-wizard-navigation-bar.horizontal ul.steps-indicator.steps-8 li{width:12.5%}aw-wizard-navigation-bar.horizontal ul.steps-indicator.steps-9:before{left:5.55555556%;right:5.55555556%}aw-wizard-navigation-bar.horizontal ul.steps-indicator.steps-9 li{width:11.11111111%}aw-wizard-navigation-bar.horizontal ul.steps-indicator.steps-10:before{left:5%;right:5%}aw-wizard-navigation-bar.horizontal ul.steps-indicator.steps-10 li{width:10%}aw-wizard-navigation-bar.horizontal ul.steps-indicator *{box-sizing:border-box}aw-wizard-navigation-bar.horizontal ul.steps-indicator li{position:relative;margin:0;padding:0;pointer-events:none;text-align:center}aw-wizard-navigation-bar.horizontal ul.steps-indicator li a{cursor:pointer}aw-wizard-navigation-bar.horizontal ul.steps-indicator li a .label{display:inline-block;padding-top:10px;color:grey;line-height:14px;font-size:14px;text-decoration:none;text-transform:uppercase;text-align:center;font-weight:700;transition:.25s}aw-wizard-navigation-bar.horizontal ul.steps-indicator li a:hover .label{color:#4d4d4d}aw-wizard-navigation-bar.horizontal ul.steps-indicator li.default a{cursor:auto}aw-wizard-navigation-bar.horizontal ul.steps-indicator li.default a .label{color:grey}aw-wizard-navigation-bar.horizontal ul.steps-indicator li.navigable{pointer-events:auto}", "aw-wizard-navigation-bar.vertical{max-width:280px;width:20%;height:100%;position:-webkit-sticky;position:sticky;top:0}aw-wizard-navigation-bar.vertical.small ul.steps-indicator{padding:5px 5px 5px 19px}aw-wizard-navigation-bar.vertical.small ul.steps-indicator li:not(:last-child):before{background-color:#e6e6e6;content:'';position:absolute;left:-7px;top:14px;height:calc(100% - 14px);width:1px}aw-wizard-navigation-bar.vertical.small ul.steps-indicator li a{min-height:14px}aw-wizard-navigation-bar.vertical.small ul.steps-indicator li .step-indicator{position:absolute;top:0;left:-14px;width:14px;height:14px;text-align:center;vertical-align:middle;line-height:14px;transition:.25s;border-radius:100%;background-color:#e6e6e6}aw-wizard-navigation-bar.vertical.small ul.steps-indicator li.current .step-indicator{background-color:grey}aw-wizard-navigation-bar.vertical.small ul.steps-indicator li.done .step-indicator{background-color:#393}aw-wizard-navigation-bar.vertical.small ul.steps-indicator li.optional .step-indicator{background-color:#38ef38}aw-wizard-navigation-bar.vertical.small ul.steps-indicator li.editing .step-indicator{background-color:red}aw-wizard-navigation-bar.vertical.small ul.steps-indicator li a:hover .step-indicator{background-color:#d9d9d9}aw-wizard-navigation-bar.vertical.small ul.steps-indicator li.current a:hover .step-indicator{background-color:#737373}aw-wizard-navigation-bar.vertical.small ul.steps-indicator li.done a:hover .step-indicator{background-color:#2d862d}aw-wizard-navigation-bar.vertical.small ul.steps-indicator li.optional a:hover .step-indicator{background-color:#20ed20}aw-wizard-navigation-bar.vertical.small ul.steps-indicator li.editing a:hover .step-indicator{background-color:#e60000}aw-wizard-navigation-bar.vertical.small ul.steps-indicator li.default a:hover .step-indicator{background-color:grey}aw-wizard-navigation-bar.vertical.large-filled ul.steps-indicator{padding:5px 5px 5px 55px}aw-wizard-navigation-bar.vertical.large-filled ul.steps-indicator li:not(:last-child):before{background-color:#e6e6e6;content:'';position:absolute;left:-25px;top:50px;height:calc(100% - 50px);width:1px}aw-wizard-navigation-bar.vertical.large-filled ul.steps-indicator li a{min-height:50px}aw-wizard-navigation-bar.vertical.large-filled ul.steps-indicator li .step-indicator{position:absolute;top:0;left:-50px;width:50px;height:50px;text-align:center;vertical-align:middle;line-height:50px;transition:.25s;border-radius:100%;background-color:#e6e6e6}aw-wizard-navigation-bar.vertical.large-filled ul.steps-indicator li.current .step-indicator{background-color:grey}aw-wizard-navigation-bar.vertical.large-filled ul.steps-indicator li.done .step-indicator{background-color:#393}aw-wizard-navigation-bar.vertical.large-filled ul.steps-indicator li.optional .step-indicator{background-color:#38ef38}aw-wizard-navigation-bar.vertical.large-filled ul.steps-indicator li.editing .step-indicator{background-color:red}aw-wizard-navigation-bar.vertical.large-filled ul.steps-indicator li a:hover .step-indicator{background-color:#d9d9d9}aw-wizard-navigation-bar.vertical.large-filled ul.steps-indicator li.current a:hover .step-indicator{background-color:#737373}aw-wizard-navigation-bar.vertical.large-filled ul.steps-indicator li.done a:hover .step-indicator{background-color:#2d862d}aw-wizard-navigation-bar.vertical.large-filled ul.steps-indicator li.optional a:hover .step-indicator{background-color:#20ed20}aw-wizard-navigation-bar.vertical.large-filled ul.steps-indicator li.editing a:hover .step-indicator{background-color:#e60000}aw-wizard-navigation-bar.vertical.large-filled ul.steps-indicator li.default a:hover .step-indicator{background-color:grey}aw-wizard-navigation-bar.vertical.large-empty ul.steps-indicator{padding:5px 5px 5px 55px}aw-wizard-navigation-bar.vertical.large-empty ul.steps-indicator li:not(:last-child):before{background-color:#e6e6e6;content:'';position:absolute;left:-25px;top:50px;height:calc(100% - 50px);width:1px}aw-wizard-navigation-bar.vertical.large-empty ul.steps-indicator li a{min-height:50px}aw-wizard-navigation-bar.vertical.large-empty ul.steps-indicator li .step-indicator{position:absolute;top:0;left:-50px;width:50px;height:50px;text-align:center;vertical-align:middle;line-height:46px;transition:.25s;border-radius:100%;border:2px solid #e6e6e6}aw-wizard-navigation-bar.vertical.large-empty ul.steps-indicator li.current .step-indicator{border:2px solid grey}aw-wizard-navigation-bar.vertical.large-empty ul.steps-indicator li.done .step-indicator{border:2px solid #393}aw-wizard-navigation-bar.vertical.large-empty ul.steps-indicator li.optional .step-indicator{border:2px solid #38ef38}aw-wizard-navigation-bar.vertical.large-empty ul.steps-indicator li.editing .step-indicator{border:2px solid red}aw-wizard-navigation-bar.vertical.large-empty ul.steps-indicator li a:hover .step-indicator{border-color:#cdcdcd}aw-wizard-navigation-bar.vertical.large-empty ul.steps-indicator li.current a:hover .step-indicator{border-color:#676767}aw-wizard-navigation-bar.vertical.large-empty ul.steps-indicator li.done a:hover .step-indicator{border-color:#267326}aw-wizard-navigation-bar.vertical.large-empty ul.steps-indicator li.optional a:hover .step-indicator{border-color:#12e212}aw-wizard-navigation-bar.vertical.large-empty ul.steps-indicator li.editing a:hover .step-indicator{border-color:#c00}aw-wizard-navigation-bar.vertical.large-empty ul.steps-indicator li.default a:hover .step-indicator{border-color:grey}aw-wizard-navigation-bar.vertical.large-filled-symbols ul.steps-indicator{padding:5px 5px 5px 55px}aw-wizard-navigation-bar.vertical.large-filled-symbols ul.steps-indicator li:not(:last-child):before{background-color:#e6e6e6;content:'';position:absolute;left:-25px;top:50px;height:calc(100% - 50px);width:1px}aw-wizard-navigation-bar.vertical.large-filled-symbols ul.steps-indicator li a{min-height:50px}aw-wizard-navigation-bar.vertical.large-filled-symbols ul.steps-indicator li .step-indicator{position:absolute;top:0;left:-50px;width:50px;height:50px;text-align:center;vertical-align:middle;line-height:50px;transition:.25s;border-radius:100%;background-color:#e6e6e6;color:#000}aw-wizard-navigation-bar.vertical.large-filled-symbols ul.steps-indicator li.current .step-indicator{background-color:grey;color:#000}aw-wizard-navigation-bar.vertical.large-filled-symbols ul.steps-indicator li.done .step-indicator{background-color:#393;color:#000}aw-wizard-navigation-bar.vertical.large-filled-symbols ul.steps-indicator li.optional .step-indicator{background-color:#38ef38;color:#000}aw-wizard-navigation-bar.vertical.large-filled-symbols ul.steps-indicator li.editing .step-indicator{background-color:red;color:#000}aw-wizard-navigation-bar.vertical.large-filled-symbols ul.steps-indicator li a:hover .step-indicator{background-color:#d9d9d9}aw-wizard-navigation-bar.vertical.large-filled-symbols ul.steps-indicator li.current a:hover .step-indicator{background-color:#737373}aw-wizard-navigation-bar.vertical.large-filled-symbols ul.steps-indicator li.done a:hover .step-indicator{background-color:#2d862d}aw-wizard-navigation-bar.vertical.large-filled-symbols ul.steps-indicator li.optional a:hover .step-indicator{background-color:#20ed20}aw-wizard-navigation-bar.vertical.large-filled-symbols ul.steps-indicator li.editing a:hover .step-indicator{background-color:#e60000}aw-wizard-navigation-bar.vertical.large-filled-symbols ul.steps-indicator li.default a:hover .step-indicator{background-color:grey}aw-wizard-navigation-bar.vertical.large-empty-symbols ul.steps-indicator{padding:5px 5px 5px 55px}aw-wizard-navigation-bar.vertical.large-empty-symbols ul.steps-indicator li:not(:last-child):before{background-color:#e6e6e6;content:'';position:absolute;left:-25px;top:50px;height:calc(100% - 50px);width:1px}aw-wizard-navigation-bar.vertical.large-empty-symbols ul.steps-indicator li a{min-height:50px}aw-wizard-navigation-bar.vertical.large-empty-symbols ul.steps-indicator li .step-indicator{position:absolute;top:0;left:-50px;width:50px;height:50px;text-align:center;vertical-align:middle;line-height:46px;transition:.25s;border-radius:100%;color:#e6e6e6;border:2px solid #e6e6e6}aw-wizard-navigation-bar.vertical.large-empty-symbols ul.steps-indicator li.current .step-indicator{color:grey;border:2px solid grey}aw-wizard-navigation-bar.vertical.large-empty-symbols ul.steps-indicator li.done .step-indicator{color:#393;border:2px solid #393}aw-wizard-navigation-bar.vertical.large-empty-symbols ul.steps-indicator li.optional .step-indicator{color:#38ef38;border:2px solid #38ef38}aw-wizard-navigation-bar.vertical.large-empty-symbols ul.steps-indicator li.editing .step-indicator{color:red;border:2px solid red}aw-wizard-navigation-bar.vertical.large-empty-symbols ul.steps-indicator li a:hover .step-indicator{border-color:#cdcdcd;color:#cdcdcd}aw-wizard-navigation-bar.vertical.large-empty-symbols ul.steps-indicator li.current a:hover .step-indicator{border-color:#676767;color:#676767}aw-wizard-navigation-bar.vertical.large-empty-symbols ul.steps-indicator li.done a:hover .step-indicator{border-color:#267326;color:#267326}aw-wizard-navigation-bar.vertical.large-empty-symbols ul.steps-indicator li.optional a:hover .step-indicator{border-color:#12e212;color:#12e212}aw-wizard-navigation-bar.vertical.large-empty-symbols ul.steps-indicator li.editing a:hover .step-indicator{border-color:#c00;color:#c00}aw-wizard-navigation-bar.vertical.large-empty-symbols ul.steps-indicator li.default a:hover .step-indicator{border-color:grey;color:grey}aw-wizard-navigation-bar.vertical ul.steps-indicator{display:flex;flex-direction:column;justify-content:center;list-style:none;margin:auto}aw-wizard-navigation-bar.vertical ul.steps-indicator *{box-sizing:border-box}aw-wizard-navigation-bar.vertical ul.steps-indicator li{position:relative;pointer-events:none}aw-wizard-navigation-bar.vertical ul.steps-indicator li:not(:last-child){margin-bottom:0;padding-bottom:10px}aw-wizard-navigation-bar.vertical ul.steps-indicator li a{display:flex;flex-direction:row;align-items:center;cursor:pointer}aw-wizard-navigation-bar.vertical ul.steps-indicator li a .label{margin-left:15px;color:grey;line-height:14px;font-size:14px;text-decoration:none;text-transform:uppercase;text-align:left;font-weight:700;transition:.25s}aw-wizard-navigation-bar.vertical ul.steps-indicator li a:hover .label{color:#4d4d4d}aw-wizard-navigation-bar.vertical ul.steps-indicator li.default a{cursor:auto}aw-wizard-navigation-bar.vertical ul.steps-indicator li.default a .label{color:grey}aw-wizard-navigation-bar.vertical ul.steps-indicator li.navigable{pointer-events:auto}"]
                    }] }
        ];
        /** @nocollapse */
        WizardNavigationBarComponent.ctorParameters = function () {
            return [
                { type: WizardState }
            ];
        };
        WizardNavigationBarComponent.propDecorators = {
            direction: [{ type: core.Input }]
        };
        return WizardNavigationBarComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `aw-wizard-step` component is used to define a normal step inside a wizard.
     *
     * ### Syntax
     *
     * With `stepTitle` and `navigationSymbol` inputs:
     *
     * ```html
     * <aw-wizard-step [stepTitle]="step title" [navigationSymbol]="{ symbol: 'symbol', fontFamily: 'font-family' }"
     *    [canExit]="deciding function" (stepEnter)="enter function" (stepExit)="exit function">
     *    ...
     * </aw-wizard-step>
     * ```
     *
     * With `awWizardStepTitle` and `awWizardStepSymbol` directives:
     *
     * ```html
     * <aw-wizard-step"
     *    [canExit]="deciding function" (stepEnter)="enter function" (stepExit)="exit function">
     *    <ng-template awWizardStepTitle>
     *        step title
     *    </ng-template>
     *    <ng-template awWizardStepSymbol>
     *        symbol
     *    </ng-template>
     *    ...
     * </aw-wizard-step>
     * ```
     *
     * ### Example
     *
     * With `stepTitle` and `navigationSymbol` inputs:
     *
     * ```html
     * <aw-wizard-step stepTitle="Address information" [navigationSymbol]="{ symbol: '&#xf1ba;', fontFamily: 'FontAwesome' }">
     *    ...
     * </aw-wizard-step>
     * ```
     *
     * With `awWizardStepTitle` and `awWizardStepSymbol` directives:
     *
     * ```html
     * <aw-wizard-step>
     *    <ng-template awWizardStepTitle>
     *        Address information
     *    </ng-template>
     *    <ng-template awWizardStepSymbol>
     *        <i class="fa fa-taxi"></i>
     *    </ng-template>
     * </aw-wizard-step>
     * ```
     *
     * @author Marc Arndt
     */
    var WizardStepComponent = /** @class */ (function (_super) {
        __extends(WizardStepComponent, _super);
        function WizardStepComponent() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        WizardStepComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'aw-wizard-step',
                        template: "<ng-content></ng-content>\n",
                        encapsulation: core.ViewEncapsulation.None,
                        providers: [
                            { provide: WizardStep, useExisting: core.forwardRef(( /**
                                     * @return {?}
                                     */function () { return WizardStepComponent; })) }
                        ],
                        styles: ["aw-wizard-step{height:auto;width:100%}"]
                    }] }
        ];
        return WizardStepComponent;
    }(WizardStep));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `aw-wizard` component defines the root component of a wizard.
     * Through the setting of input parameters for the `aw-wizard` component it's possible to change the location and size
     * of its navigation bar.
     *
     * ### Syntax
     * ```html
     * <aw-wizard [navBarLocation]="location of navigation bar" [navBarLayout]="layout of navigation bar">
     *     ...
     * </aw-wizard>
     * ```
     *
     * ### Example
     *
     * Without completion step:
     *
     * ```html
     * <aw-wizard navBarLocation="top" navBarLayout="small">
     *     <aw-wizard-step>...</aw-wizard-step>
     *     <aw-wizard-step>...</aw-wizard-step>
     * </aw-wizard>
     * ```
     *
     * With completion step:
     *
     * ```html
     * <aw-wizard navBarLocation="top" navBarLayout="small">
     *     <aw-wizard-step>...</aw-wizard-step>
     *     <aw-wizard-step>...</aw-wizard-step>
     *     <aw-wizard-completion-step>...</aw-wizard-completion-step>
     * </aw-wizard>
     * ```
     *
     * @author Marc Arndt
     */
    var WizardComponent = /** @class */ (function () {
        /**
         * Constructor
         *
         * @param model The model for this wizard component
         */
        function WizardComponent(model) {
            this.model = model;
            /**
             * The location of the navigation bar inside the wizard.
             * This location can be either top, bottom, left or right
             */
            this.navBarLocation = 'top';
            /**
             * The layout of the navigation bar inside the wizard.
             * The layout can be either small, large-filled, large-empty or large-symbols
             */
            this.navBarLayout = 'small';
            /**
             * The direction in which the steps inside the navigation bar should be shown.
             * The direction can be either `left-to-right` or `right-to-left`
             */
            this.navBarDirection = 'left-to-right';
            /**
             * The navigation mode used for transitioning between different steps.
             * The navigation mode can be either `strict`, `semi-strict` or `free`
             */
            this.navigationMode = 'strict';
            /**
             * The initially selected step, represented by its index
             */
            this.defaultStepIndex = 0;
            /**
             * True, if the navigation bar shouldn't be used for navigating
             */
            this.disableNavigationBar = false;
        }
        Object.defineProperty(WizardComponent.prototype, "horizontalOrientation", {
            /**
             * Returns true if this wizard uses a horizontal orientation.
             * The wizard uses a horizontal orientation, iff the navigation bar is shown at the top or bottom of this wizard
             *
             * @returns True if this wizard uses a horizontal orientation
             */
            get: /**
             * Returns true if this wizard uses a horizontal orientation.
             * The wizard uses a horizontal orientation, iff the navigation bar is shown at the top or bottom of this wizard
             *
             * @return {?} True if this wizard uses a horizontal orientation
             */ function () {
                return this.navBarLocation === 'top' || this.navBarLocation === 'bottom';
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(WizardComponent.prototype, "verticalOrientation", {
            /**
             * Returns true if this wizard uses a vertical orientation.
             * The wizard uses a vertical orientation, iff the navigation bar is shown at the left or right of this wizard
             *
             * @returns True if this wizard uses a vertical orientation
             */
            get: /**
             * Returns true if this wizard uses a vertical orientation.
             * The wizard uses a vertical orientation, iff the navigation bar is shown at the left or right of this wizard
             *
             * @return {?} True if this wizard uses a vertical orientation
             */ function () {
                return this.navBarLocation === 'left' || this.navBarLocation === 'right';
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(WizardComponent.prototype, "navigation", {
            /**
             * The navigation mode for this wizard
             */
            get: /**
             * The navigation mode for this wizard
             * @return {?}
             */ function () {
                return this.model.navigationMode;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * Updates the model after certain input values have changed
         *
         * @param changes The detected changes
         */
        /**
         * Updates the model after certain input values have changed
         *
         * @param {?} changes The detected changes
         * @return {?}
         */
        WizardComponent.prototype.ngOnChanges = /**
         * Updates the model after certain input values have changed
         *
         * @param {?} changes The detected changes
         * @return {?}
         */
            function (changes) {
                var e_1, _a;
                try {
                    for (var _b = __values(Object.keys(changes)), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var propName = _c.value;
                        /** @type {?} */
                        var change = changes[propName];
                        if (!change.firstChange) {
                            switch (propName) {
                                case 'defaultStepIndex':
                                    this.model.defaultStepIndex = parseInt(change.currentValue, 10);
                                    break;
                                case 'disableNavigationBar':
                                    this.model.disableNavigationBar = change.currentValue;
                                    break;
                                case 'navigationMode':
                                    this.model.updateNavigationMode(change.currentValue);
                                    break;
                                /* istanbul ignore next */
                                default:
                            }
                        }
                    }
                }
                catch (e_1_1) {
                    e_1 = { error: e_1_1 };
                }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return))
                            _a.call(_b);
                    }
                    finally {
                        if (e_1)
                            throw e_1.error;
                    }
                }
            };
        /**
         * Initialization work
         */
        /**
         * Initialization work
         * @return {?}
         */
        WizardComponent.prototype.ngAfterContentInit = /**
         * Initialization work
         * @return {?}
         */
            function () {
                var _this = this;
                // add a subscriber to the wizard steps QueryList to listen to changes in the DOM
                this.wizardSteps.changes.subscribe(( /**
                 * @param {?} changedWizardSteps
                 * @return {?}
                 */function (changedWizardSteps) {
                    _this.model.updateWizardSteps(changedWizardSteps.toArray());
                }));
                // initialize the model
                this.model.disableNavigationBar = this.disableNavigationBar;
                this.model.defaultStepIndex = this.defaultStepIndex;
                this.model.updateWizardSteps(this.wizardSteps.toArray());
                this.model.updateNavigationMode(this.navigationMode);
                // finally reset the whole wizard state
                this.navigation.reset();
            };
        WizardComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'aw-wizard',
                        template: "<aw-wizard-navigation-bar\n  [direction]=\"navBarDirection\"\n  *ngIf=\"navBarLocation == 'top' || navBarLocation == 'left'\"\n  [ngClass]=\"{\n    vertical: navBarLocation == 'left',\n    horizontal: navBarLocation == 'top',\n    small: navBarLayout == 'small',\n    'large-filled': navBarLayout == 'large-filled',\n    'large-filled-symbols': navBarLayout == 'large-filled-symbols',\n    'large-empty': navBarLayout == 'large-empty',\n    'large-empty-symbols': navBarLayout == 'large-empty-symbols'\n  }\">\n</aw-wizard-navigation-bar>\n\n<div [ngClass]=\"{\n  'wizard-steps': true,\n  vertical: navBarLocation == 'left' || navBarLocation == 'right',\n  horizontal: navBarLocation == 'top' || navBarLocation == 'bottom'\n}\">\n  <ng-content></ng-content>\n</div>\n\n<aw-wizard-navigation-bar\n  [direction]=\"navBarDirection\"\n  *ngIf=\"navBarLocation == 'bottom' || navBarLocation == 'right'\"\n  [ngClass]=\"{\n    vertical: navBarLocation == 'right',\n    horizontal: navBarLocation == 'bottom',\n    small: navBarLayout == 'small',\n    'large-filled': navBarLayout == 'large-filled',\n    'large-filled-symbols': navBarLayout == 'large-filled-symbols',\n    'large-empty': navBarLayout == 'large-empty',\n    'large-empty-symbols': navBarLayout == 'large-empty-symbols'\n  }\">\n</aw-wizard-navigation-bar>\n",
                        encapsulation: core.ViewEncapsulation.None,
                        providers: [WizardState],
                        styles: ["aw-wizard{display:flex;justify-content:flex-start}aw-wizard.vertical{flex-direction:row}aw-wizard.horizontal{flex-direction:column}aw-wizard .wizard-steps{top:0;display:flex}aw-wizard .wizard-steps.vertical{min-width:calc(100% - 280px);width:80%;height:100%;flex-direction:column}aw-wizard .wizard-steps.horizontal{width:100%;flex-direction:row}"]
                    }] }
        ];
        /** @nocollapse */
        WizardComponent.ctorParameters = function () {
            return [
                { type: WizardState }
            ];
        };
        WizardComponent.propDecorators = {
            wizardSteps: [{ type: core.ContentChildren, args: [WizardStep,] }],
            navBarLocation: [{ type: core.Input }],
            navBarLayout: [{ type: core.Input }],
            navBarDirection: [{ type: core.Input }],
            navigationMode: [{ type: core.Input }],
            defaultStepIndex: [{ type: core.Input }],
            disableNavigationBar: [{ type: core.Input }],
            horizontalOrientation: [{ type: core.HostBinding, args: ['class.horizontal',] }],
            verticalOrientation: [{ type: core.HostBinding, args: ['class.vertical',] }]
        };
        return WizardComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `awEnableBackLinks` directive can be used to allow the user to leave a [[WizardCompletionStep]] after is has been entered.
     *
     * ### Syntax
     *
     * ```html
     * <aw-wizard-completion-step awEnableBackLinks (stepExit)="exit function">
     *     ...
     * </aw-wizard-completion-step>
     * ```
     *
     * ### Example
     *
     * ```html
     * <aw-wizard-completion-step stepTitle="Final step" awEnableBackLinks>
     *     ...
     * </aw-wizard-completion-step>
     * ```
     *
     * @author Marc Arndt
     */
    var EnableBackLinksDirective = /** @class */ (function () {
        /**
         * Constructor
         *
         * @param completionStep The wizard completion step, which should be exitable
         */
        function EnableBackLinksDirective(completionStep) {
            this.completionStep = completionStep;
            /**
             * This EventEmitter is called when the step is exited.
             * The bound method can be used to do cleanup work.
             */
            this.stepExit = new core.EventEmitter();
        }
        /**
         * Initialization work
         */
        /**
         * Initialization work
         * @return {?}
         */
        EnableBackLinksDirective.prototype.ngOnInit = /**
         * Initialization work
         * @return {?}
         */
            function () {
                this.completionStep.canExit = true;
                this.completionStep.stepExit = this.stepExit;
            };
        EnableBackLinksDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[awEnableBackLinks]'
                    },] }
        ];
        /** @nocollapse */
        EnableBackLinksDirective.ctorParameters = function () {
            return [
                { type: WizardCompletionStep, decorators: [{ type: core.Host }] }
            ];
        };
        EnableBackLinksDirective.propDecorators = {
            stepExit: [{ type: core.Output }]
        };
        return EnableBackLinksDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * Checks whether the given `value` implements the interface [[StepId]].
     *
     * @param {?} value The value to be checked
     * @return {?} True if the given value implements [[StepId]] and false otherwise
     */
    function isStepId(value) {
        return value.hasOwnProperty('stepId') && !(value instanceof WizardStep);
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * Checks whether the given `value` implements the interface [[StepIndex]].
     *
     * @param {?} value The value to be checked
     * @return {?} True if the given value implements [[StepIndex]] and false otherwise
     */
    function isStepIndex(value) {
        return value.hasOwnProperty('stepIndex');
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * Checks whether the given `value` implements the interface [[StepOffset]].
     *
     * @param {?} value The value to be checked
     * @return {?} True if the given value implements [[StepOffset]] and false otherwise
     */
    function isStepOffset(value) {
        return value.hasOwnProperty('stepOffset');
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `awGoToStep` directive can be used to navigate to a given step.
     * This step can be defined in one of multiple formats
     *
     * ### Syntax
     *
     * With absolute step index:
     *
     * ```html
     * <button [awGoToStep]="{ stepIndex: absolute step index }" (finalize)="finalize method">...</button>
     * ```
     *
     * With unique step id:
     *
     * ```html
     * <button [awGoToStep]="{ stepId: 'step id of destination step' }" (finalize)="finalize method">...</button>
     * ```
     *
     * With a wizard step object:
     *
     * ```html
     * <button [awGoToStep]="wizard step object" (finalize)="finalize method">...</button>
     * ```
     *
     * With an offset to the defining step:
     *
     * ```html
     * <button [awGoToStep]="{ stepOffset: offset }" (finalize)="finalize method">...</button>
     * ```
     *
     * @author Marc Arndt
     */
    var GoToStepDirective = /** @class */ (function () {
        /**
         * Constructor
         *
         * @param wizardState The wizard state
         * @param wizardStep The wizard step, which contains this [[GoToStepDirective]]
         */
        function GoToStepDirective(wizardState, wizardStep) {
            this.wizardState = wizardState;
            this.wizardStep = wizardStep;
            /**
             * This [[EventEmitter]] is called directly before the current step is exited during a transition through a component with this directive.
             */
            this.preFinalize = new core.EventEmitter();
            /**
             * This [[EventEmitter]] is called directly after the current step is exited during a transition through a component with this directive.
             */
            this.postFinalize = new core.EventEmitter();
        }
        Object.defineProperty(GoToStepDirective.prototype, "finalize", {
            /**
             * A convenience field for `preFinalize`
             */
            get: /**
             * A convenience field for `preFinalize`
             * @return {?}
             */ function () {
                return this.preFinalize;
            },
            /**
             * A convenience name for `preFinalize`
             *
             * @param emitter The [[EventEmitter]] to be set
             */
            set: /**
             * A convenience name for `preFinalize`
             *
             * @param {?} emitter The [[EventEmitter]] to be set
             * @return {?}
             */ function (emitter) {
                /* istanbul ignore next */
                this.preFinalize = emitter;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(GoToStepDirective.prototype, "destinationStep", {
            /**
             * Returns the destination step of this directive as an absolute step index inside the wizard
             *
             * @returns The index of the destination step
             * @throws If `targetStep` is of an unknown type an `Error` is thrown
             */
            get: /**
             * Returns the destination step of this directive as an absolute step index inside the wizard
             *
             * @throws If `targetStep` is of an unknown type an `Error` is thrown
             * @return {?} The index of the destination step
             */ function () {
                /** @type {?} */
                var destinationStep;
                if (isStepIndex(this.targetStep)) {
                    destinationStep = this.targetStep.stepIndex;
                }
                else if (isStepId(this.targetStep)) {
                    destinationStep = this.wizardState.getIndexOfStepWithId(this.targetStep.stepId);
                }
                else if (isStepOffset(this.targetStep) && this.wizardStep !== null) {
                    destinationStep = this.wizardState.getIndexOfStep(this.wizardStep) + this.targetStep.stepOffset;
                }
                else if (this.targetStep instanceof WizardStep) {
                    destinationStep = this.wizardState.getIndexOfStep(this.targetStep);
                }
                else {
                    throw new Error("Input 'targetStep' is neither a WizardStep, StepOffset, StepIndex or StepId");
                }
                return destinationStep;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(GoToStepDirective.prototype, "navigationMode", {
            /**
             * The navigation mode
             */
            get: /**
             * The navigation mode
             * @private
             * @return {?}
             */ function () {
                return this.wizardState.navigationMode;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * Listener method for `click` events on the component with this directive.
         * After this method is called the wizard will try to transition to the `destinationStep`
         */
        /**
         * Listener method for `click` events on the component with this directive.
         * After this method is called the wizard will try to transition to the `destinationStep`
         * @param {?} event
         * @return {?}
         */
        GoToStepDirective.prototype.onClick = /**
         * Listener method for `click` events on the component with this directive.
         * After this method is called the wizard will try to transition to the `destinationStep`
         * @param {?} event
         * @return {?}
         */
            function (event) {
                this.navigationMode.goToStep(this.destinationStep, this.preFinalize, this.postFinalize);
            };
        GoToStepDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[awGoToStep]'
                    },] }
        ];
        /** @nocollapse */
        GoToStepDirective.ctorParameters = function () {
            return [
                { type: WizardState },
                { type: WizardStep, decorators: [{ type: core.Optional }] }
            ];
        };
        GoToStepDirective.propDecorators = {
            preFinalize: [{ type: core.Output }],
            postFinalize: [{ type: core.Output }],
            targetStep: [{ type: core.Input, args: ['awGoToStep',] }],
            finalize: [{ type: core.Output }],
            onClick: [{ type: core.HostListener, args: ['click', ['$event'],] }]
        };
        return GoToStepDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `awNextStep` directive can be used to navigate to the next step.
     *
     * ### Syntax
     *
     * ```html
     * <button awNextStep (finalize)="finalize method">...</button>
     * ```
     *
     * @author Marc Arndt
     */
    var NextStepDirective = /** @class */ (function () {
        /**
         * Constructor
         *
         * @param wizardState The state of the wizard
         */
        function NextStepDirective(wizardState) {
            this.wizardState = wizardState;
            /**
             * This [[EventEmitter]] is called directly before the current step is exited during a transition through a component with this directive.
             */
            this.preFinalize = new core.EventEmitter();
            /**
             * This [[EventEmitter]] is called directly after the current step is exited during a transition through a component with this directive.
             */
            this.postFinalize = new core.EventEmitter();
        }
        Object.defineProperty(NextStepDirective.prototype, "finalize", {
            /**
             * A convenience field for `preFinalize`
             */
            get: /**
             * A convenience field for `preFinalize`
             * @return {?}
             */ function () {
                return this.preFinalize;
            },
            /**
             * A convenience name for `preFinalize`
             *
             * @param emitter The [[EventEmitter]] to be set
             */
            set: /**
             * A convenience name for `preFinalize`
             *
             * @param {?} emitter The [[EventEmitter]] to be set
             * @return {?}
             */ function (emitter) {
                /* istanbul ignore next */
                this.preFinalize = emitter;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(NextStepDirective.prototype, "navigationMode", {
            /**
             * The navigation mode
             */
            get: /**
             * The navigation mode
             * @private
             * @return {?}
             */ function () {
                return this.wizardState.navigationMode;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * Listener method for `click` events on the component with this directive.
         * After this method is called the wizard will try to transition to the next step
         */
        /**
         * Listener method for `click` events on the component with this directive.
         * After this method is called the wizard will try to transition to the next step
         * @param {?} event
         * @return {?}
         */
        NextStepDirective.prototype.onClick = /**
         * Listener method for `click` events on the component with this directive.
         * After this method is called the wizard will try to transition to the next step
         * @param {?} event
         * @return {?}
         */
            function (event) {
                this.navigationMode.goToNextStep(this.preFinalize, this.postFinalize);
            };
        NextStepDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[awNextStep]'
                    },] }
        ];
        /** @nocollapse */
        NextStepDirective.ctorParameters = function () {
            return [
                { type: WizardState }
            ];
        };
        NextStepDirective.propDecorators = {
            preFinalize: [{ type: core.Output }],
            postFinalize: [{ type: core.Output }],
            finalize: [{ type: core.Output }],
            onClick: [{ type: core.HostListener, args: ['click', ['$event'],] }]
        };
        return NextStepDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `awOptionalStep` directive can be used to define an optional `wizard-step`.
     * An optional wizard step is a [[WizardStep]] that doesn't need to be completed to transition to later wizard steps.
     *
     * ### Syntax
     *
     * ```html
     * <aw-wizard-step awOptionalStep>
     *     ...
     * </aw-wizard-step>
     * ```
     *
     * ### Example
     *
     * ```html
     * <aw-wizard-step stepTitle="Second step" awOptionalStep>
     *     ...
     * </aw-wizard-step>
     * ```
     *
     * @author Marc Arndt
     */
    var OptionalStepDirective = /** @class */ (function () {
        /**
         * Constructor
         *
         * @param wizardStep The wizard step, which contains this [[OptionalStepDirective]]
         */
        function OptionalStepDirective(wizardStep) {
            this.wizardStep = wizardStep;
        }
        /**
         * Initialization work
         */
        /**
         * Initialization work
         * @return {?}
         */
        OptionalStepDirective.prototype.ngOnInit = /**
         * Initialization work
         * @return {?}
         */
            function () {
                this.wizardStep.optional = true;
            };
        OptionalStepDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[awOptionalStep]'
                    },] }
        ];
        /** @nocollapse */
        OptionalStepDirective.ctorParameters = function () {
            return [
                { type: WizardStep, decorators: [{ type: core.Host }] }
            ];
        };
        return OptionalStepDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `awPreviousStep` directive can be used to navigate to the previous step.
     * Compared to the [[NextStepDirective]] it's important to note, that this directive doesn't contain a `finalize` output method.
     *
     * ### Syntax
     *
     * ```html
     * <button awPreviousStep>...</button>
     * ```
     *
     * @author Marc Arndt
     */
    var PreviousStepDirective = /** @class */ (function () {
        /**
         * Constructor
         *
         * @param wizardState The state of the wizard
         */
        function PreviousStepDirective(wizardState) {
            this.wizardState = wizardState;
            /**
             * This [[EventEmitter]] is called directly before the current step is exited during a transition through a component with this directive.
             */
            this.preFinalize = new core.EventEmitter();
            /**
             * This [[EventEmitter]] is called directly after the current step is exited during a transition through a component with this directive.
             */
            this.postFinalize = new core.EventEmitter();
        }
        Object.defineProperty(PreviousStepDirective.prototype, "finalize", {
            /**
             * A convenience field for `preFinalize`
             */
            get: /**
             * A convenience field for `preFinalize`
             * @return {?}
             */ function () {
                return this.preFinalize;
            },
            /**
             * A convenience field for `preFinalize`
             *
             * @param emitter The [[EventEmitter]] to be set
             */
            set: /**
             * A convenience field for `preFinalize`
             *
             * @param {?} emitter The [[EventEmitter]] to be set
             * @return {?}
             */ function (emitter) {
                /* istanbul ignore next */
                this.preFinalize = emitter;
            },
            enumerable: true,
            configurable: true
        });
        Object.defineProperty(PreviousStepDirective.prototype, "navigationMode", {
            /**
             * The navigation mode
             */
            get: /**
             * The navigation mode
             * @private
             * @return {?}
             */ function () {
                return this.wizardState.navigationMode;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * Listener method for `click` events on the component with this directive.
         * After this method is called the wizard will try to transition to the previous step
         */
        /**
         * Listener method for `click` events on the component with this directive.
         * After this method is called the wizard will try to transition to the previous step
         * @param {?} event
         * @return {?}
         */
        PreviousStepDirective.prototype.onClick = /**
         * Listener method for `click` events on the component with this directive.
         * After this method is called the wizard will try to transition to the previous step
         * @param {?} event
         * @return {?}
         */
            function (event) {
                this.navigationMode.goToPreviousStep(this.preFinalize, this.postFinalize);
            };
        PreviousStepDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[awPreviousStep]'
                    },] }
        ];
        /** @nocollapse */
        PreviousStepDirective.ctorParameters = function () {
            return [
                { type: WizardState }
            ];
        };
        PreviousStepDirective.propDecorators = {
            preFinalize: [{ type: core.Output }],
            postFinalize: [{ type: core.Output }],
            finalize: [{ type: core.Output }],
            onClick: [{ type: core.HostListener, args: ['click', ['$event'],] }]
        };
        return PreviousStepDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `awResetWizard` directive can be used to reset the wizard to its initial state.
     * This directive accepts an output, which can be used to specify some custom cleanup work during the reset process.
     *
     * ### Syntax
     *
     * ```html
     * <button awResetWizard (finalize)="custom reset task">...</button>
     * ```
     *
     * @author Marc Arndt
     */
    var ResetWizardDirective = /** @class */ (function () {
        /**
         * Constructor
         *
         * @param wizardState The wizard state
         */
        function ResetWizardDirective(wizardState) {
            this.wizardState = wizardState;
            /**
             * An [[EventEmitter]] containing some tasks to be done, directly before the wizard is being reset
             */
            this.finalize = new core.EventEmitter();
        }
        Object.defineProperty(ResetWizardDirective.prototype, "navigationMode", {
            /**
             * The navigation mode
             */
            get: /**
             * The navigation mode
             * @private
             * @return {?}
             */ function () {
                return this.wizardState.navigationMode;
            },
            enumerable: true,
            configurable: true
        });
        /**
         * Resets the wizard
         */
        /**
         * Resets the wizard
         * @param {?} event
         * @return {?}
         */
        ResetWizardDirective.prototype.onClick = /**
         * Resets the wizard
         * @param {?} event
         * @return {?}
         */
            function (event) {
                // do some optional cleanup work
                this.finalize.emit();
                // reset the wizard to its initial state
                this.navigationMode.reset();
            };
        ResetWizardDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[awResetWizard]'
                    },] }
        ];
        /** @nocollapse */
        ResetWizardDirective.ctorParameters = function () {
            return [
                { type: WizardState }
            ];
        };
        ResetWizardDirective.propDecorators = {
            finalize: [{ type: core.Output }],
            onClick: [{ type: core.HostListener, args: ['click', ['$event'],] }]
        };
        return ResetWizardDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `awSelectedStep` directive can be used on a [[WizardStep]] to set it as selected after the wizard initialisation or a reset.
     *
     * ### Syntax
     *
     * ```html
     * <aw-wizard-step stepTitle="Step title" awSelectedStep>
     *     ...
     * </aw-wizard-step>
     * ```
     *
     * @author Marc Arndt
     */
    var SelectedStepDirective = /** @class */ (function () {
        /**
         * Constructor
         *
         * @param wizardStep The wizard step, which should be selected by default
         */
        function SelectedStepDirective(wizardStep) {
            this.wizardStep = wizardStep;
        }
        /**
         * Initialization work
         */
        /**
         * Initialization work
         * @return {?}
         */
        SelectedStepDirective.prototype.ngOnInit = /**
         * Initialization work
         * @return {?}
         */
            function () {
                this.wizardStep.defaultSelected = true;
            };
        SelectedStepDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[awSelectedStep]'
                    },] }
        ];
        /** @nocollapse */
        SelectedStepDirective.ctorParameters = function () {
            return [
                { type: WizardStep, decorators: [{ type: core.Host }] }
            ];
        };
        return SelectedStepDirective;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `awWizardCompletionStep` directive can be used to define a completion/success step at the end of your wizard
     * After a [[WizardCompletionStep]] has been entered, it has the characteristic that the user is blocked from
     * leaving it again to a previous step.
     * In addition entering a [[WizardCompletionStep]] automatically sets the `wizard`, and all steps inside the `wizard`,
     * as completed.
     *
     * ### Syntax
     *
     * ```html
     * <div awWizardCompletionStep [stepTitle]="title of the wizard step"
     *    [navigationSymbol]="{ symbol: 'navigation symbol', fontFamily: 'font-family' }"
     *    (stepEnter)="event emitter to be called when the wizard step is entered"
     *    (stepExit)="event emitter to be called when the wizard step is exited">
     *    ...
     * </div>
     * ```
     *
     * ### Example
     *
     * ```html
     * <div awWizardCompletionStep stepTitle="Step 1" [navigationSymbol]="{ symbol: '1' }">
     *    ...
     * </div>
     * ```
     *
     * With a navigation symbol from the `font-awesome` font:
     *
     * ```html
     * <div awWizardCompletionStep stepTitle="Step 1" [navigationSymbol]="{ symbol: '&#xf1ba;', fontFamily: 'FontAwesome' }">
     *    ...
     * </div>
     * ```
     *
     * @author Marc Arndt
     */
    var WizardCompletionStepDirective = /** @class */ (function (_super) {
        __extends(WizardCompletionStepDirective, _super);
        function WizardCompletionStepDirective() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        WizardCompletionStepDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[awWizardCompletionStep]',
                        providers: [
                            { provide: WizardStep, useExisting: core.forwardRef(( /**
                                     * @return {?}
                                     */function () { return WizardCompletionStepDirective; })) },
                            { provide: WizardCompletionStep, useExisting: core.forwardRef(( /**
                                     * @return {?}
                                     */function () { return WizardCompletionStepDirective; })) }
                        ]
                    },] }
        ];
        return WizardCompletionStepDirective;
    }(WizardCompletionStep));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The `awWizardStep` directive can be used to define a normal step inside a wizard.
     *
     * ### Syntax
     *
     * With `stepTitle` and `navigationSymbol` inputs:
     *
     * ```html
     * <div awWizardStep [stepTitle]="step title" [navigationSymbol]="{ symbol: 'symbol', fontFamily: 'font-family' }"
     *    [canExit]="deciding function" (stepEnter)="enter function" (stepExit)="exit function">
     *    ...
     * </div>
     * ```
     *
     * With `awWizardStepTitle` and `awWizardStepSymbol` directives:
     *
     * ```html
     * <div awWizardStep [canExit]="deciding function" (stepEnter)="enter function" (stepExit)="exit function">
     *    <ng-template awWizardStepTitle>
     *        step title
     *    </ng-template>
     *    <ng-template awWizardStepSymbol>
     *        symbol
     *    </ng-template>
     *    ...
     * </div>
     * ```
     *
     * ### Example
     *
     * With `stepTitle` and `navigationSymbol` inputs:
     *
     * ```html
     * <div awWizardStep stepTitle="Address information" [navigationSymbol]="{ symbol: '&#xf1ba;', fontFamily: 'FontAwesome' }">
     *    ...
     * </div>
     * ```
     *
     * With `awWizardStepTitle` and `awWizardStepSymbol` directives:
     *
     * ```html
     * <div awWizardStep>
     *    <ng-template awWizardStepTitle>
     *        Address information
     *    </ng-template>
     *    <ng-template awWizardStepSymbol>
     *        <i class="fa fa-taxi"></i>
     *    </ng-template>
     * </div>
     * ```
     *
     * @author Marc Arndt
     */
    var WizardStepDirective = /** @class */ (function (_super) {
        __extends(WizardStepDirective, _super);
        function WizardStepDirective() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        WizardStepDirective.decorators = [
            { type: core.Directive, args: [{
                        selector: '[awWizardStep]',
                        providers: [
                            { provide: WizardStep, useExisting: core.forwardRef(( /**
                                     * @return {?}
                                     */function () { return WizardStepDirective; })) }
                        ]
                    },] }
        ];
        return WizardStepDirective;
    }(WizardStep));

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */
    /**
     * The module defining all the content inside `angular-archwizard`
     *
     * @author Marc Arndt
     */
    var ArchwizardModule = /** @class */ (function () {
        function ArchwizardModule() {
        }
        /* istanbul ignore next */
        /* istanbul ignore next */
        /**
         * @return {?}
         */
        ArchwizardModule.forRoot = /* istanbul ignore next */
            /**
             * @return {?}
             */
            function () {
                return { ngModule: ArchwizardModule, providers: [] };
            };
        ArchwizardModule.decorators = [
            { type: core.NgModule, args: [{
                        declarations: [
                            WizardComponent,
                            WizardStepComponent,
                            WizardNavigationBarComponent,
                            WizardCompletionStepComponent,
                            GoToStepDirective,
                            NextStepDirective,
                            PreviousStepDirective,
                            OptionalStepDirective,
                            WizardStepSymbolDirective,
                            WizardStepTitleDirective,
                            EnableBackLinksDirective,
                            WizardStepDirective,
                            WizardCompletionStepDirective,
                            SelectedStepDirective,
                            ResetWizardDirective
                        ],
                        imports: [
                            common.CommonModule
                        ],
                        exports: [
                            WizardComponent,
                            WizardStepComponent,
                            WizardNavigationBarComponent,
                            WizardCompletionStepComponent,
                            GoToStepDirective,
                            NextStepDirective,
                            PreviousStepDirective,
                            OptionalStepDirective,
                            WizardStepSymbolDirective,
                            WizardStepTitleDirective,
                            EnableBackLinksDirective,
                            WizardStepDirective,
                            WizardCompletionStepDirective,
                            SelectedStepDirective,
                            ResetWizardDirective
                        ]
                    },] }
        ];
        return ArchwizardModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
     */

    exports.WizardCompletionStepComponent = WizardCompletionStepComponent;
    exports.WizardNavigationBarComponent = WizardNavigationBarComponent;
    exports.WizardStepComponent = WizardStepComponent;
    exports.WizardComponent = WizardComponent;
    exports.EnableBackLinksDirective = EnableBackLinksDirective;
    exports.GoToStepDirective = GoToStepDirective;
    exports.NextStepDirective = NextStepDirective;
    exports.OptionalStepDirective = OptionalStepDirective;
    exports.PreviousStepDirective = PreviousStepDirective;
    exports.ResetWizardDirective = ResetWizardDirective;
    exports.SelectedStepDirective = SelectedStepDirective;
    exports.WizardCompletionStepDirective = WizardCompletionStepDirective;
    exports.WizardStepDirective = WizardStepDirective;
    exports.WizardStepTitleDirective = WizardStepTitleDirective;
    exports.FreeNavigationMode = FreeNavigationMode;
    exports.NavigationMode = NavigationMode;
    exports.SemiStrictNavigationMode = SemiStrictNavigationMode;
    exports.StrictNavigationMode = StrictNavigationMode;
    exports.WizardState = WizardState;
    exports.navigationModeFactory = navigationModeFactory;
    exports.MovingDirection = MovingDirection;
    exports.isStepId = isStepId;
    exports.isStepIndex = isStepIndex;
    exports.isStepOffset = isStepOffset;
    exports.WizardCompletionStep = WizardCompletionStep;
    exports.WizardStep = WizardStep;
    exports.ArchwizardModule = ArchwizardModule;
    exports.a = WizardStepSymbolDirective;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
