// ngx-chartist/bundles/ngx-chartist.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('chartist'), require('@angular/common')) :
    typeof define === 'function' && define.amd ? define('ngx-chartist', ['exports', '@angular/core', 'chartist', '@angular/common'], factory) :
    (factory((global['ngx-chartist'] = {}),global.ng.core,null,global.ng.common));
}(this, (function (exports,i0,Chartist,common) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    function __values(o) {
        var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
        if (m)
            return m.call(o);
        return {
            next: function () {
                if (o && i >= o.length)
                    o = void 0;
                return { value: o && o[i++], done: !o };
            }
        };
    }
    function __read(o, n) {
        var m = typeof Symbol === "function" && o[Symbol.iterator];
        if (!m)
            return o;
        var i = m.call(o), r, ar = [], e;
        try {
            while ((n === void 0 || n-- > 0) && !(r = i.next()).done)
                ar.push(r.value);
        }
        catch (error) {
            e = { error: error };
        }
        finally {
            try {
                if (r && !r.done && (m = i["return"]))
                    m.call(i);
            }
            finally {
                if (e)
                    throw e.error;
            }
        }
        return ar;
    }
    function __spread() {
        for (var ar = [], i = 0; i < arguments.length; i++)
            ar = ar.concat(__read(arguments[i]));
        return ar;
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var NgxChartistService = /** @class */ (function () {
        function NgxChartistService() {
        }
        /**
         * @return {?}
         */
        NgxChartistService.prototype.getPrecision = /**
         * @return {?}
         */
            function () {
                return Chartist.precision;
            };
        /**
         * @return {?}
         */
        NgxChartistService.prototype.getEscapingMap = /**
         * @return {?}
         */
            function () {
                return Chartist.escapingMap;
            };
        /**
         * @return {?}
         */
        NgxChartistService.prototype.getPie = /**
         * @return {?}
         */
            function () {
                return Chartist.Pie;
            };
        /**
         * @return {?}
         */
        NgxChartistService.prototype.getBar = /**
         * @return {?}
         */
            function () {
                return Chartist.Bar;
            };
        /**
         * @return {?}
         */
        NgxChartistService.prototype.getLine = /**
         * @return {?}
         */
            function () {
                return Chartist.Line;
            };
        /**
         * @return {?}
         */
        NgxChartistService.prototype.getCandle = /**
         * @return {?}
         */
            function () {
                return Chartist.Candle;
            };
        /**
         * @return {?}
         */
        NgxChartistService.prototype.getFixedScaleAxis = /**
         * @return {?}
         */
            function () {
                return Chartist.FixedScaleAxis;
            };
        /**
         * @return {?}
         */
        NgxChartistService.prototype.getAutoScaleAxis = /**
         * @return {?}
         */
            function () {
                return Chartist.AutoScaleAxis;
            };
        /**
         * @return {?}
         */
        NgxChartistService.prototype.getStepAxis = /**
         * @return {?}
         */
            function () {
                return Chartist.StepAxis;
            };
        /**
         * @return {?}
         */
        NgxChartistService.prototype.getSvg = /**
         * @return {?}
         */
            function () {
                return Chartist.Svg;
            };
        /**
         * @return {?}
         */
        NgxChartistService.prototype.getInterpolation = /**
         * @return {?}
         */
            function () {
                return Chartist.Interpolation;
            };
        /**
         * @return {?}
         */
        NgxChartistService.prototype.getNoop = /**
         * @return {?}
         */
            function () {
                return Chartist.noop;
            };
        /**
         * @return {?}
         */
        NgxChartistService.prototype.getPlugins = /**
         * @return {?}
         */
            function () {
                return Chartist.plugins;
            };
        /**
         * @param {?} n
         * @return {?}
         */
        NgxChartistService.prototype.alphaNumerate = /**
         * @param {?} n
         * @return {?}
         */
            function (n) {
                return Chartist.alphaNumerate(n);
            };
        /**
         * @param {?} target
         * @param {...?} sources
         * @return {?}
         */
        NgxChartistService.prototype.extend = /**
         * @param {?} target
         * @param {...?} sources
         * @return {?}
         */
            function (target) {
                var sources = [];
                for (var _i = 1; _i < arguments.length; _i++) {
                    sources[_i - 1] = arguments[_i];
                }
                return Chartist.extend.apply(Chartist, __spread([target], sources));
            };
        /**
         * @param {?} str
         * @param {?} subStr
         * @param {?} newSubStr
         * @return {?}
         */
        NgxChartistService.prototype.replaceAll = /**
         * @param {?} str
         * @param {?} subStr
         * @param {?} newSubStr
         * @return {?}
         */
            function (str, subStr, newSubStr) {
                return Chartist.replaceAll(str, subStr, newSubStr);
            };
        /**
         * @param {?} value
         * @param {?} unit
         * @return {?}
         */
        NgxChartistService.prototype.ensureUnit = /**
         * @param {?} value
         * @param {?} unit
         * @return {?}
         */
            function (value, unit) {
                return Chartist.ensureUnit(value, unit);
            };
        /**
         * @param {?} input
         * @return {?}
         */
        NgxChartistService.prototype.quantity = /**
         * @param {?} input
         * @return {?}
         */
            function (input) {
                return Chartist.quantity(input);
            };
        /**
         * @param {?} query
         * @return {?}
         */
        NgxChartistService.prototype.query = /**
         * @param {?} query
         * @return {?}
         */
            function (query) {
                return Chartist.query(query);
            };
        /**
         * @param {?} length
         * @return {?}
         */
        NgxChartistService.prototype.times = /**
         * @param {?} length
         * @return {?}
         */
            function (length) {
                return Chartist.times(length);
            };
        /**
         * @param {?} previous
         * @param {?} current
         * @return {?}
         */
        NgxChartistService.prototype.sum = /**
         * @param {?} previous
         * @param {?} current
         * @return {?}
         */
            function (previous, current) {
                return Chartist.sum(previous, current);
            };
        /**
         * @param {?} factor
         * @return {?}
         */
        NgxChartistService.prototype.mapMultiply = /**
         * @param {?} factor
         * @return {?}
         */
            function (factor) {
                return Chartist.mapMultiply(factor);
            };
        /**
         * @param {?} addend
         * @return {?}
         */
        NgxChartistService.prototype.mapAdd = /**
         * @param {?} addend
         * @return {?}
         */
            function (addend) {
                return Chartist.mapAdd(addend);
            };
        /**
         * @param {?} arr
         * @param {?} cb
         * @return {?}
         */
        NgxChartistService.prototype.serialMap = /**
         * @param {?} arr
         * @param {?} cb
         * @return {?}
         */
            function (arr, cb) {
                return Chartist.serialMap(arr, cb);
            };
        /**
         * @param {?} value
         * @param {?=} digits
         * @return {?}
         */
        NgxChartistService.prototype.roundWithPrecision = /**
         * @param {?} value
         * @param {?=} digits
         * @return {?}
         */
            function (value, digits) {
                return Chartist.roundWithPrecision(value, digits);
            };
        /**
         * @param {?} value
         * @param {?=} dimension
         * @return {?}
         */
        NgxChartistService.prototype.getMultiValue = /**
         * @param {?} value
         * @param {?=} dimension
         * @return {?}
         */
            function (value, dimension) {
                return Chartist.getMultiValue(value, dimension);
            };
        /**
         * @param {?} data
         * @return {?}
         */
        NgxChartistService.prototype.serialize = /**
         * @param {?} data
         * @return {?}
         */
            function (data) {
                return Chartist.serialize(data);
            };
        /**
         * @param {?} data
         * @return {?}
         */
        NgxChartistService.prototype.deserialize = /**
         * @param {?} data
         * @return {?}
         */
            function (data) {
                return Chartist.deserialize(data);
            };
        /**
         * @param {?} container
         * @param {?} width
         * @param {?} height
         * @param {?} className
         * @return {?}
         */
        NgxChartistService.prototype.createSvg = /**
         * @param {?} container
         * @param {?} width
         * @param {?} height
         * @param {?} className
         * @return {?}
         */
            function (container, width, height, className) {
                return Chartist.createSvg(container, width, height, className);
            };
        NgxChartistService.decorators = [
            { type: i0.Injectable, args: [{
                        providedIn: 'root'
                    },] }
        ];
        NgxChartistService.ctorParameters = function () { return []; };
        /** @nocollapse */ NgxChartistService.ngInjectableDef = i0.defineInjectable({ factory: function NgxChartistService_Factory() { return new NgxChartistService(); }, token: NgxChartistService, providedIn: "root" });
        return NgxChartistService;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var NgxChartistComponent = /** @class */ (function () {
        function NgxChartistComponent(element) {
            this.element = element.nativeElement;
        }
        /**
         * @return {?}
         */
        NgxChartistComponent.prototype.ngOnInit = /**
         * @return {?}
         */
            function () {
                var _this = this;
                if (!this.type || !this.data) {
                    Promise.reject('Expected at least type and data.');
                }
                return this.renderChart().then(function (chart) {
                    if (_this.events !== undefined) {
                        _this.bindEvents(chart);
                    }
                    return chart;
                });
            };
        /**
         * @param {?} changes
         * @return {?}
         */
        NgxChartistComponent.prototype.ngOnChanges = /**
         * @param {?} changes
         * @return {?}
         */
            function (changes) {
                if (this.element) {
                    this.update(changes);
                }
            };
        /**
         * @return {?}
         */
        NgxChartistComponent.prototype.ngOnDestroy = /**
         * @return {?}
         */
            function () {
                if (this.chart) {
                    this.chart.detach();
                }
            };
        /**
         * @return {?}
         */
        NgxChartistComponent.prototype.renderChart = /**
         * @return {?}
         */
            function () {
                var _this = this;
                /** @type {?} */
                var promises = [
                    this.type,
                    this.element,
                    this.data,
                    this.options,
                    this.responsiveOptions
                ];
                if (this.lineSmooth) {
                    switch (this.lineSmooth['interpolation']) {
                        case 'cardinal': {
                            this.options['lineSmooth'] = Chartist.Interpolation.cardinal(this.lineSmooth['data']);
                            break;
                        }
                        case 'simple': {
                            this.options['lineSmooth'] = Chartist.Interpolation.simple(this.lineSmooth['data']);
                            break;
                        }
                        case 'none': {
                            this.options['lineSmooth'] = Chartist.Interpolation.none(this.lineSmooth['data']);
                            break;
                        }
                        case 'step': {
                            this.options['lineSmooth'] = Chartist.Interpolation.step(this.lineSmooth['data']);
                            break;
                        }
                    }
                }
                return Promise.all(promises).then(function (values) {
                    var _a = __read(values), type = _a[0], args = _a.slice(1);
                    if (!(type in Chartist)) {
                        throw new Error(type + " is not a valid chart type");
                    }
                    _this.chart = (( /** @type {?} */(Chartist)))[type].apply((( /** @type {?} */(Chartist))), __spread(args));
                    return _this.chart;
                });
            };
        /**
         * @param {?} changes
         * @return {?}
         */
        NgxChartistComponent.prototype.update = /**
         * @param {?} changes
         * @return {?}
         */
            function (changes) {
                if (!this.chart || 'type' in changes) {
                    this.renderChart();
                }
                else {
                    if (changes.data) {
                        this.data = changes.data.currentValue;
                    }
                    if (changes.options) {
                        this.options = changes.options.currentValue;
                    }
                    (( /** @type {?} */(this.chart))).update(this.data, this.options);
                }
            };
        /**
         * @param {?} chart
         * @return {?}
         */
        NgxChartistComponent.prototype.bindEvents = /**
         * @param {?} chart
         * @return {?}
         */
            function (chart) {
                var e_1, _a;
                try {
                    for (var _b = __values(Object.keys(this.events)), _c = _b.next(); !_c.done; _c = _b.next()) {
                        var event_1 = _c.value;
                        chart.on(event_1, this.events[event_1]);
                    }
                }
                catch (e_1_1) {
                    e_1 = { error: e_1_1 };
                }
                finally {
                    try {
                        if (_c && !_c.done && (_a = _b.return))
                            _a.call(_b);
                    }
                    finally {
                        if (e_1)
                            throw e_1.error;
                    }
                }
            };
        NgxChartistComponent.decorators = [
            { type: i0.Component, args: [{
                        selector: 'lib-ngx-chartist',
                        template: '',
                        styles: [":host:not(.without-styles) ::ng-deep .ct-label{fill:rgba(0,0,0,.4);color:rgba(0,0,0,.4);font-size:.75rem;line-height:1}:host:not(.without-styles) ::ng-deep .ct-chart-bar .ct-label,:host:not(.without-styles) ::ng-deep .ct-chart-line .ct-label{display:block;display:flex}:host:not(.without-styles) ::ng-deep .ct-chart-donut .ct-label,:host:not(.without-styles) ::ng-deep .ct-chart-pie .ct-label{dominant-baseline:central}:host:not(.without-styles) ::ng-deep .ct-label.ct-horizontal.ct-start{align-items:flex-end;justify-content:flex-start;text-align:left;text-anchor:start}:host:not(.without-styles) ::ng-deep .ct-label.ct-horizontal.ct-end{align-items:flex-start;justify-content:flex-start;text-align:left;text-anchor:start}:host:not(.without-styles) ::ng-deep .ct-label.ct-vertical.ct-start{align-items:flex-end;justify-content:flex-end;text-align:right;text-anchor:end}:host:not(.without-styles) ::ng-deep .ct-label.ct-vertical.ct-end{align-items:flex-end;justify-content:flex-start;text-align:left;text-anchor:start}:host:not(.without-styles) ::ng-deep .ct-chart-bar .ct-label.ct-horizontal.ct-start{align-items:flex-end;justify-content:center;text-align:center;text-anchor:start}:host:not(.without-styles) ::ng-deep .ct-chart-bar .ct-label.ct-horizontal.ct-end{align-items:flex-start;justify-content:center;text-align:center;text-anchor:start}:host:not(.without-styles) ::ng-deep .ct-chart-bar.ct-horizontal-bars .ct-label.ct-horizontal.ct-start{align-items:flex-end;justify-content:flex-start;text-align:left;text-anchor:start}:host:not(.without-styles) ::ng-deep .ct-chart-bar.ct-horizontal-bars .ct-label.ct-horizontal.ct-end{align-items:flex-start;justify-content:flex-start;text-align:left;text-anchor:start}:host:not(.without-styles) ::ng-deep .ct-chart-bar.ct-horizontal-bars .ct-label.ct-vertical.ct-start{align-items:center;justify-content:flex-end;text-align:right;text-anchor:end}:host:not(.without-styles) ::ng-deep .ct-chart-bar.ct-horizontal-bars .ct-label.ct-vertical.ct-end{align-items:center;justify-content:flex-start;text-align:left;text-anchor:end}:host:not(.without-styles) ::ng-deep .ct-grid{stroke:rgba(0,0,0,.2);stroke-width:1px;stroke-dasharray:2px}:host:not(.without-styles) ::ng-deep .ct-grid-background{fill:none}:host:not(.without-styles) ::ng-deep .ct-point{stroke-width:10px;stroke-linecap:round}:host:not(.without-styles) ::ng-deep .ct-line{fill:none;stroke-width:4px}:host:not(.without-styles) ::ng-deep .ct-area{stroke:none;fill-opacity:.1}:host:not(.without-styles) ::ng-deep .ct-bar{fill:none;stroke-width:10px}:host:not(.without-styles) ::ng-deep .ct-slice-donut{fill:none;stroke-width:60px}:host:not(.without-styles) ::ng-deep .ct-series-a .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-a .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-a .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-a .ct-slice-donut{stroke:#d70206}:host:not(.without-styles) ::ng-deep .ct-series-a .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-a .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-a .ct-slice-pie{fill:#d70206}:host:not(.without-styles) ::ng-deep .ct-series-b .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-b .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-b .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-b .ct-slice-donut{stroke:#f05b4f}:host:not(.without-styles) ::ng-deep .ct-series-b .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-b .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-b .ct-slice-pie{fill:#f05b4f}:host:not(.without-styles) ::ng-deep .ct-series-c .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-c .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-c .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-c .ct-slice-donut{stroke:#f4c63d}:host:not(.without-styles) ::ng-deep .ct-series-c .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-c .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-c .ct-slice-pie{fill:#f4c63d}:host:not(.without-styles) ::ng-deep .ct-series-d .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-d .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-d .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-d .ct-slice-donut{stroke:#d17905}:host:not(.without-styles) ::ng-deep .ct-series-d .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-d .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-d .ct-slice-pie{fill:#d17905}:host:not(.without-styles) ::ng-deep .ct-series-e .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-e .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-e .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-e .ct-slice-donut{stroke:#453d3f}:host:not(.without-styles) ::ng-deep .ct-series-e .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-e .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-e .ct-slice-pie{fill:#453d3f}:host:not(.without-styles) ::ng-deep .ct-series-f .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-f .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-f .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-f .ct-slice-donut{stroke:#59922b}:host:not(.without-styles) ::ng-deep .ct-series-f .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-f .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-f .ct-slice-pie{fill:#59922b}:host:not(.without-styles) ::ng-deep .ct-series-g .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-g .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-g .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-g .ct-slice-donut{stroke:#0544d3}:host:not(.without-styles) ::ng-deep .ct-series-g .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-g .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-g .ct-slice-pie{fill:#0544d3}:host:not(.without-styles) ::ng-deep .ct-series-h .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-h .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-h .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-h .ct-slice-donut{stroke:#6b0392}:host:not(.without-styles) ::ng-deep .ct-series-h .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-h .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-h .ct-slice-pie{fill:#6b0392}:host:not(.without-styles) ::ng-deep .ct-series-i .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-i .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-i .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-i .ct-slice-donut{stroke:#f05b4f}:host:not(.without-styles) ::ng-deep .ct-series-i .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-i .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-i .ct-slice-pie{fill:#f05b4f}:host:not(.without-styles) ::ng-deep .ct-series-j .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-j .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-j .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-j .ct-slice-donut{stroke:#dda458}:host:not(.without-styles) ::ng-deep .ct-series-j .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-j .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-j .ct-slice-pie{fill:#dda458}:host:not(.without-styles) ::ng-deep .ct-series-k .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-k .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-k .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-k .ct-slice-donut{stroke:#eacf7d}:host:not(.without-styles) ::ng-deep .ct-series-k .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-k .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-k .ct-slice-pie{fill:#eacf7d}:host:not(.without-styles) ::ng-deep .ct-series-l .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-l .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-l .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-l .ct-slice-donut{stroke:#86797d}:host:not(.without-styles) ::ng-deep .ct-series-l .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-l .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-l .ct-slice-pie{fill:#86797d}:host:not(.without-styles) ::ng-deep .ct-series-m .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-m .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-m .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-m .ct-slice-donut{stroke:#b2c326}:host:not(.without-styles) ::ng-deep .ct-series-m .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-m .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-m .ct-slice-pie{fill:#b2c326}:host:not(.without-styles) ::ng-deep .ct-series-n .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-n .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-n .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-n .ct-slice-donut{stroke:#6188e2}:host:not(.without-styles) ::ng-deep .ct-series-n .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-n .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-n .ct-slice-pie{fill:#6188e2}:host:not(.without-styles) ::ng-deep .ct-series-o .ct-bar,:host:not(.without-styles) ::ng-deep .ct-series-o .ct-line,:host:not(.without-styles) ::ng-deep .ct-series-o .ct-point,:host:not(.without-styles) ::ng-deep .ct-series-o .ct-slice-donut{stroke:#a748ca}:host:not(.without-styles) ::ng-deep .ct-series-o .ct-area,:host:not(.without-styles) ::ng-deep .ct-series-o .ct-slice-donut-solid,:host:not(.without-styles) ::ng-deep .ct-series-o .ct-slice-pie{fill:#a748ca}:host:not(.without-styles) ::ng-deep .ct-square{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-square:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:100%}:host:not(.without-styles) ::ng-deep .ct-square:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-square>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-minor-second{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-minor-second:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:93.75%}:host:not(.without-styles) ::ng-deep .ct-minor-second:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-minor-second>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-major-second{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-major-second:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:88.88889%}:host:not(.without-styles) ::ng-deep .ct-major-second:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-major-second>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-minor-third{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-minor-third:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:83.33333%}:host:not(.without-styles) ::ng-deep .ct-minor-third:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-minor-third>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-major-third{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-major-third:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:80%}:host:not(.without-styles) ::ng-deep .ct-major-third:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-major-third>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-perfect-fourth{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-perfect-fourth:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:75%}:host:not(.without-styles) ::ng-deep .ct-perfect-fourth:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-perfect-fourth>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-perfect-fifth{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-perfect-fifth:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:66.66667%}:host:not(.without-styles) ::ng-deep .ct-perfect-fifth:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-perfect-fifth>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-minor-sixth{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-minor-sixth:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:62.5%}:host:not(.without-styles) ::ng-deep .ct-minor-sixth:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-minor-sixth>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-golden-section{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-golden-section:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:61.8047%}:host:not(.without-styles) ::ng-deep .ct-golden-section:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-golden-section>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-major-sixth{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-major-sixth:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:60%}:host:not(.without-styles) ::ng-deep .ct-major-sixth:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-major-sixth>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-minor-seventh{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-minor-seventh:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:56.25%}:host:not(.without-styles) ::ng-deep .ct-minor-seventh:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-minor-seventh>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-major-seventh{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-major-seventh:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:53.33333%}:host:not(.without-styles) ::ng-deep .ct-major-seventh:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-major-seventh>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-octave{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-octave:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:50%}:host:not(.without-styles) ::ng-deep .ct-octave:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-octave>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-major-tenth{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-major-tenth:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:40%}:host:not(.without-styles) ::ng-deep .ct-major-tenth:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-major-tenth>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-major-eleventh{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-major-eleventh:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:37.5%}:host:not(.without-styles) ::ng-deep .ct-major-eleventh:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-major-eleventh>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-major-twelfth{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-major-twelfth:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:33.33333%}:host:not(.without-styles) ::ng-deep .ct-major-twelfth:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-major-twelfth>svg{display:block;position:absolute;top:0;left:0}:host:not(.without-styles) ::ng-deep .ct-double-octave{display:block;position:relative;width:100%}:host:not(.without-styles) ::ng-deep .ct-double-octave:before{display:block;float:left;content:\"\";width:0;height:0;padding-bottom:25%}:host:not(.without-styles) ::ng-deep .ct-double-octave:after{content:\"\";display:table;clear:both}:host:not(.without-styles) ::ng-deep .ct-double-octave>svg{display:block;position:absolute;top:0;left:0}"]
                    }] }
        ];
        NgxChartistComponent.ctorParameters = function () {
            return [
                { type: i0.ElementRef }
            ];
        };
        NgxChartistComponent.propDecorators = {
            data: [{ type: i0.Input }],
            type: [{ type: i0.Input }],
            options: [{ type: i0.Input }],
            lineSmooth: [{ type: i0.Input }],
            responsiveOptions: [{ type: i0.Input }],
            events: [{ type: i0.Input }]
        };
        return NgxChartistComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    var NgxChartistModule = /** @class */ (function () {
        function NgxChartistModule() {
        }
        NgxChartistModule.decorators = [
            { type: i0.NgModule, args: [{
                        imports: [
                            common.CommonModule
                        ],
                        declarations: [NgxChartistComponent],
                        exports: [NgxChartistComponent],
                        providers: [
                            NgxChartistService
                        ]
                    },] }
        ];
        return NgxChartistModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    exports.NgxChartistService = NgxChartistService;
    exports.NgxChartistComponent = NgxChartistComponent;
    exports.NgxChartistModule = NgxChartistModule;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
